<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Denovo motif discovery of target regions using DREME — runDreme • memes</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Denovo motif discovery of target regions using DREME — runDreme" />
<meta property="og:description" content="DREME discovers short, ungapped, de-novo motifs that are relatively
enriched relative to a control set of sequences. DREME can be run to discover
motifs relative to a shuffled set of input sequences, or against a separately
provided set of &quot;control&quot; sequences." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



<!-- dependencies from examples -->


  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">memes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/install_guide.html">Install MEME</a>
    </li>
    <li>
      <a href="../articles/core_dreme.html">De-Novo Motif Discovery</a>
    </li>
    <li>
      <a href="../articles/core_tomtom.html">Motif Comparison</a>
    </li>
    <li>
      <a href="../articles/core_ame.html">Motif Enrichment</a>
    </li>
    <li>
      <a href="../articles/core_fimo.html">Motif Scanning</a>
    </li>
    <li>
      <a href="../articles/tidy_motifs.html">Motif Manipulation</a>
    </li>
    <li>
      <a href="../articles/integrative_analysis.html">ChIP-Seq Data Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/snystrom/memes/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Denovo motif discovery of target regions using DREME</h1>
    <small class="dont-index">Source: <a href='https://github.com/snystrom/memes/blob/master/R/generics.R'><code>R/generics.R</code></a></small>
    <div class="hidden name"><code>runDreme.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>DREME discovers short, ungapped, <em>de-novo</em> motifs that are relatively
enriched relative to a control set of sequences. DREME can be run to discover
motifs relative to a shuffled set of input sequences, or against a separately
provided set of "control" sequences.</p>
    </div>

    <pre class="usage"><span class='fu'>runDreme</span><span class='op'>(</span><span class='va'>input</span>, <span class='va'>control</span>, outdir <span class='op'>=</span> <span class='st'>"auto"</span>, meme_path <span class='op'>=</span> <span class='cn'>NULL</span>, silent <span class='op'>=</span> <span class='cn'>TRUE</span>, <span class='va'>...</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>input</th>
      <td><p>regions to scan for motifs. Can be any of:</p><ul>
<li><p>path to fasta file</p></li>
<li><p>DNAStringSet object (can be generated from GRanges using <code><a href='get_sequence.html'>get_sequence()</a></code>)</p></li>
<li><p>List of DNAStringSet objects (generated from <code><a href='get_sequence.html'>get_sequence()</a></code>)</p></li>
<li><p><em>NOTE:</em> if using StringSet inputs, each entry must be named (set with <code><a href='https://rdrr.io/r/base/names.html'>names()</a></code>).</p></li>
<li><p><em>NOTE:</em> If you want to retain the raw dreme output files, you must use a
path to fasta file as input, or specify an "outdir"</p></li>
</ul></td>
    </tr>
    <tr>
      <th>control</th>
      <td><p>regions to use as background for motif search. Can be any of:</p><ul>
<li><p>path to fasta file</p></li>
<li><p>DNAStringSet object (can be generated from GRanges using get_sequence)</p></li>
<li><p>A Biostrings::BStringSetList (generated using <code>get_sequence</code>), in which
case all sequences in the list will be combined as the control set.</p></li>
<li><p>if <code>input</code> is a list of DNAStringSet objects, a character vector of names
in <code>input</code> will use those sequences as background. runDreme will not scan
those regions as input.</p></li>
<li><p>"shuffle" to use dreme's built-in dinucleotide shuffle feature (NOTE: if
<code>input</code> is a list object with an entry named "shuffle", the list entry will
be used instead).
Optionally can also pass <code>seed = &lt;any number&gt;</code> to <code>...</code> to use as the random
seed during shuffling. If no seed is passed, dreme will use 1 as the random
seed, so results will be reproducible if rerunning. <strong>NOTE:</strong> beware
system-specific differences. As of v5, dreme will compile using the default
python installation on a system (either python2.7 or python3). The random
number generator changed between python2.7 and python3, so results will not
be reproducible between systems using python2.7 vs 3.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>outdir</th>
      <td><p>path to output directory of dreme files, or "auto" to autogenerate path. Default: location of
input fasta in dir named "\&lt;input\&gt;<em>vs</em>\&lt;control\&gt;". If input is
DNAstringset, will be temporary path. This means that if you want to save
the raw output files, you must use fasta files as input or use an
informative (and unique) outdir name. memes will <strong>not check</strong> if it
overwrites files in a directory. Directories will be recursively created if needed.</p></td>
    </tr>
    <tr>
      <th>meme_path</th>
      <td><p>optional, path to "meme/bin/" on your local machine.
runDreme will search 3 places in order for meme if this flag is unset:</p><ol>
<li><p>the option "meme_bin" (set with options(meme_bin = "path/to/meme/bin"))</p></li>
<li><p>the environment variable "MEME_PATH" (set in .Renviron)</p></li>
<li><p>"~/meme/bin/" as the default location</p></li>
</ol>
<ul>
<li><p>If the user sets meme_path in the function call, this value will always be used</p></li>
</ul></td>
    </tr>
    <tr>
      <th>silent</th>
      <td><p>whether to suppress printing dreme stdout as a message when
finishing with no errors. Can be useful for troubleshooting in situations
where no motifs are discovered, but command completes successfully.
(default: TRUE)</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>dreme flags can be passed as R function arguments to use
non-default behavior. For a full list of valid arguments, run your local
install of dreme -h, or visit the dreme documentation
<a href='http://meme-suite.org/doc/dreme.html'>website</a>. See list below for aliases
of common flags. To set flags with no values (ex. <code>-dna</code>), pass the
argument as a boolean value (ex. <code>dna = TRUE</code>).</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>data.frame with statistics for each discovered motif. The <code>motif</code>
column contains a universalmotif object representation in PCM format of
each DREME motif. If no motifs are discovered, returns NULL. The column
values are as follows:</p><ul>
<li><p>rank = ranked order of discovered motif</p></li>
<li><p>name = primary name of motif</p></li>
<li><p>altname = alternative name of motif</p></li>
<li><p>seq = consensus sequence of the motif</p></li>
<li><p>length = length of discovered motif</p></li>
<li><p>nsites = number of times the motif is found in input sequences</p></li>
<li><p>positive_hits = number of sequences in input containing at least 1 of the motif</p></li>
<li><p>negative_hits = number of sequences in control containing at least 1 of the motif</p></li>
<li><p>pval = p-value</p></li>
<li><p>eval = E-value</p></li>
<li><p>unerased_eval = Unerased E-Value</p></li>
<li><p>positive_total = number of sequences in input</p></li>
<li><p>negative_total = number of sequences in control</p></li>
<li><p>pos_frac = fraction of positive sequences with a hit</p></li>
<li><p>neg_frac = fraction of negative sequences with a hit</p></li>
<li><p>motif = a universalmotif object of the discovered motif</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Properly setting the <code>control</code> parameter is key to discovering biologically
relevant motifs. Often, using <code>control = "shuffle"</code> will produce a suboptimal
set of motifs; however, some discriminative analysis designs don't have
proper "control" regions other than to shuffle.</p>
<p>In addition to allowing any valid flag of dreme to be passed to <code>...</code>, we
provide a few user-friendly aliases for common flags which are more readable (see list below).
For example, e = 1 will use a max evalue cutoff of 1. This is equivalent to
setting evalue = 1. For additional details about each DREME flag, see the
<a href='http://meme-suite.org/doc/dreme.html'>DREME Manual Webpage</a>.</p>
<p>List of values which can be passed to <code>...</code>:
<strong>NOTE:</strong> values must be referred to using their name in the "memes alias"
column, not the "DREME Flag" column.</p><table class='table'>
<tr><td>memes alias</td><td>DREME Flag</td><td>description</td><td>default</td></tr>
<tr><td>nmotifs</td><td>m</td><td>max number of motifs to discover</td><td>NULL</td></tr>
<tr><td>sec</td><td>t</td><td>max number of seconds to run</td><td>NULL</td></tr>
<tr><td>evalue</td><td>e</td><td>max E-value cutoff</td><td>0.05</td></tr>
<tr><td>seed</td><td>s</td><td>random seed if using "shuffle" as control</td><td>1</td></tr>
<tr><td>ngen</td><td>g</td><td>nuber of REs to generalize</td><td>100</td></tr>
<tr><td>mink</td><td>mink</td><td>minimum motif width to search</td><td>3</td></tr>
<tr><td>maxk</td><td>maxk</td><td>maximum motif width to search</td><td>7</td></tr>
<tr><td>k</td><td>k</td><td>set mink and maxk to this value</td><td>NULL</td></tr>
<tr><td>norc</td><td>norc</td><td>search only the input strand for sequences</td><td>FALSE</td></tr>
<tr><td>dna</td><td>dna</td><td>use DNA alphabet</td><td>TRUE</td></tr>
<tr><td>rna</td><td>rna</td><td>use RNA alphabet</td><td>FALSE</td></tr>
<tr><td>protein</td><td>protein</td><td>use protein alphabet (NOT RECCOMENDED)</td><td>FALSE</td></tr>
</table>


    <h2 class="hasAnchor" id="citation"><a class="anchor" href="#citation"></a>Citation</h2>

    <p>If you use <code>runDreme()</code> in your analysis, please cite:</p>
<p>Timothy L. Bailey, "DREME: Motif discovery in transcription factor ChIP-seq
data", Bioinformatics, 27(12):1653-1659, 2011.
<a href='https://academic.oup.com/bioinformatics/article/27/12/1653/257754'>full text</a></p><h3 class='hasAnchor' id='licensing'><a class='anchor' href='#licensing'></a>Licensing</h3>


<p>The MEME Suite is free for non-profit use, but for-profit users should purchase a
license. See the <a href='http://meme-suite.org/doc/copyright.html'>MEME Suite Copyright Page</a> for details.</p>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='meme_is_installed.html'>meme_is_installed</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'># Create random named sequences as input for example</span>
<span class='va'>seqs</span> <span class='op'>&lt;-</span> <span class='fu'>universalmotif</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/universalmotif/man/create_sequences.html'>create_sequences</a></span><span class='op'>(</span>rng.seed <span class='op'>=</span> <span class='fl'>123</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>seqs</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>seqs</span><span class='op'>)</span>

<span class='co'># Runs dreme with default settings, shuffles input as background</span>
<span class='fu'>runDreme</span><span class='op'>(</span><span class='va'>seqs</span>, <span class='st'>"shuffle"</span><span class='op'>)</span>

<span class='co'># Runs searching for max 2 motifs, e-value cutoff = 0.1, explicitly using the DNA alphabet</span>
<span class='fu'>runDreme</span><span class='op'>(</span><span class='va'>seqs</span>, <span class='st'>"shuffle"</span>, nmotifs <span class='op'>=</span> <span class='fl'>2</span>, e <span class='op'>=</span> <span class='fl'>0.1</span>, dna <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; NULL</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Spencer Nystrom.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


