<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Find instances of motifs using FIMO — runFimo • memes</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Find instances of motifs using FIMO — runFimo" />
<meta property="og:description" content="FIMO scans input sequences to identify the positions of matches to each input
motif. FIMO has no sequence length or motif number restrictions." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



<!-- dependencies from examples -->


  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">memes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/install_guide.html">Install MEME</a>
    </li>
    <li>
      <a href="../articles/core_dreme.html">De-Novo Motif Discovery</a>
    </li>
    <li>
      <a href="../articles/core_tomtom.html">Motif Comparison</a>
    </li>
    <li>
      <a href="../articles/core_ame.html">Motif Enrichment</a>
    </li>
    <li>
      <a href="../articles/core_fimo.html">Motif Scanning</a>
    </li>
    <li>
      <a href="../articles/tidy_motifs.html">Motif Manipulation</a>
    </li>
    <li>
      <a href="../articles/integrative_analysis.html">ChIP-Seq Data Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/snystrom/memes/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find instances of motifs using FIMO</h1>
    <small class="dont-index">Source: <a href='https://github.com/snystrom/memes/blob/master/R/fimo.R'><code>R/fimo.R</code></a></small>
    <div class="hidden name"><code>runFimo.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>FIMO scans input sequences to identify the positions of matches to each input
motif. FIMO has no sequence length or motif number restrictions.</p>
    </div>

    <pre class="usage"><span class='fu'>runFimo</span><span class='op'>(</span>
  <span class='va'>sequences</span>,
  <span class='va'>motifs</span>,
  bfile <span class='op'>=</span> <span class='st'>"motif"</span>,
  outdir <span class='op'>=</span> <span class='st'>"auto"</span>,
  parse_genomic_coord <span class='op'>=</span> <span class='cn'>TRUE</span>,
  skip_matched_sequence <span class='op'>=</span> <span class='cn'>FALSE</span>,
  max_strand <span class='op'>=</span> <span class='cn'>TRUE</span>,
  text <span class='op'>=</span> <span class='cn'>TRUE</span>,
  meme_path <span class='op'>=</span> <span class='cn'>NULL</span>,
  silent <span class='op'>=</span> <span class='cn'>TRUE</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>sequences</th>
      <td><p>path to fasta file, or stringset input.</p></td>
    </tr>
    <tr>
      <th>motifs</th>
      <td><p>path to .meme format file, or universalmotif/universalmotif list input.</p></td>
    </tr>
    <tr>
      <th>bfile</th>
      <td><p>path to background file, or special values: "motif" to use
0-order frequencies contained in the motif, or "uniform" to use a uniform
letter distribution. (default: "motif")</p></td>
    </tr>
    <tr>
      <th>outdir</th>
      <td><p>output directory location. Only used if text = FALSE. Default:
"auto" to autogenerate directory name. Note: if not using a fasta path as
input, this will be a temporary location unless explicity set.</p></td>
    </tr>
    <tr>
      <th>parse_genomic_coord</th>
      <td><p><code>logical(1)</code> whether to parse genomic position
from fasta headers. Fasta headers must be UCSC format positions (ie
"chr:start-end"), but base 1 indexed (GRanges format). If names of fasta
entries are genomic coordinates and parse_genomic_coord == TRUE, results
will contain genomic coordinates of motif matches, otherwise FIMO will return
relative coordinates (i.e. positions from 1 to length of the fasta entry).</p></td>
    </tr>
    <tr>
      <th>skip_matched_sequence</th>
      <td><p><code>logical(1)</code> whether or not to include the DNA
sequence of the match. Default: <code>FALSE</code>. Note: jobs will complete faster if
set to <code>TRUE</code>. <code><a href='add_sequence.html'>add_sequence()</a></code> can be used to lookup the sequence after data import if
<code>parse_genomic_coord</code> is <code>TRUE</code>, so setting this flag is not strictly needed.</p></td>
    </tr>
    <tr>
      <th>max_strand</th>
      <td><p>if match is found on both strands, only report strand with
best match (default: TRUE).</p></td>
    </tr>
    <tr>
      <th>text</th>
      <td><p><code>logical(1)</code> (default: <code>TRUE</code>). No output files will be created
on the filesystem. The results are unsorted and no q-values are computed.
This setting allows fast searches on very large inputs. When set to <code>FALSE</code>
FIMO will discard 50% of the lower significance matches if &gt;100,000 matches are
detected. <code>text = FALSE</code> will also incur a performance penalty because it
must first read a file to disk, then read it into memory. For these reasons,
I suggest keeping <code>text = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>meme_path</th>
      <td><p>path to <code>meme/bin/</code> (optional). Defaut: <code>NULL</code>, searches
"MEME_PATH" environment variable or "meme_path" option for path to "meme/bin/".</p></td>
    </tr>
    <tr>
      <th>silent</th>
      <td><p><code>logical(1)</code> whether to suppress stdout/stderr printing to
console (default: TRUE). If the command is failing or giving unexpected
output, setting <code>silent = FALSE</code> can aid troubleshooting.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional commandline arguments to fimo. See the FIMO Flag table below.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>GRanges object containing positions of each match. Note: if
<code>parse_genomic_coords = FALSE</code>, each <code>seqnames</code> entry will be the full fasta
header, and start/end will be the relative position within that sequence of the
match. The GRanges object has the following additional <code>mcols</code>:
* motif_id = primary name of matched motif
* motif_alt_id = alternate name of matched motif
* score = score of match (higher score is a better match)
* pvalue = pvalue of the match
* qvalue = qvalue of the match
* matched_sequence = sequence that matches the motif</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Additional arguments passed to <code>...</code>. See: <a href='http://meme-suite.org/doc/fimo.html?man_type=web'>Fimo web manual</a>
for a complete description of FIMO flags.</p><table class='table'>
<tr><td>FIMO Flag</td><td>allowed values</td><td>default</td><td>description</td></tr>
<tr><td>alpha</td><td><code>numeric(1)</code></td><td>1</td><td>alpha for calculating position-specific priors. Represents fraction of sites that are binding sites of TF of interest. Used in conjunction with <code>psp</code></td></tr>
<tr><td>bfile</td><td>"motif", "motif-file", "uniform", file path,</td><td>"motif"</td><td>If "motif" or "motif-file", use 0-order letter frequencies from motif. "uniform" sets uniform letter frequencies.</td></tr>
<tr><td>max_stored_scores</td><td><code>integer(1)</code></td><td>NULL</td><td>maximum number of scores to be stored for computing q-values. used when <code>text = FALSE</code>, see FIMO webpage for details</td></tr>
<tr><td>motif_pseudo</td><td><code>numeric(1)</code></td><td>0.1</td><td>pseudocount added to motif matrix</td></tr>
<tr><td>no_qvalue</td><td><code>logical(1)</code></td><td>FALSE</td><td>only needed when <code>text = FALSE</code>, do not compute q-value for each p-value</td></tr>
<tr><td>norc</td><td><code>logical(1)</code></td><td>FALSE</td><td>Do not score reverse complement strand</td></tr>
<tr><td>prior_dist</td><td>file path</td><td>NULL</td><td>file containing binned distribution of priors</td></tr>
<tr><td>psp</td><td>file path</td><td>NULL</td><td>file containing position specific priors. Requires <code>prior_dist</code></td></tr>
<tr><td>qv_thresh</td><td><code>logical(1)</code></td><td>FALSE</td><td>use q-values for the output threshold</td></tr>
<tr><td>thresh</td><td><code>numeric(1)</code></td><td><code>1e-4</code></td><td>output threshold for returning a match, only matches with values less than <code>thresh</code> are returned.</td></tr>
</table>


<h3 class='hasAnchor' id='licensing'><a class='anchor' aria-hidden='true' href='#licensing'></a>Licensing</h3>


<p>The MEME Suite is free for non-profit use, but for-profit users should purchase a
license. See the <a href='http://meme-suite.org/doc/copyright.html'>MEME Suite Copyright Page</a> for details.</p>

    <h2 class="hasAnchor" id="citation"><a class="anchor" href="#citation"></a>Citation</h2>

    <p>If you use <code>runFimo()</code> in your analysis, please cite:</p>
<p>Charles E. Grant, Timothy L. Bailey, and William Stafford Noble, "FIMO:
Scanning for occurrences of a given motif", Bioinformatics, 27(7):1017-1018,
2011. <a href='http://bioinformatics.oxfordjournals.org/content/early/2011/02/16/bioinformatics.btr064.full'>full text</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='meme_is_installed.html'>meme_is_installed</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span><span class='op'>{</span>
<span class='co'># Generate some example input sequences</span>
<span class='va'>seq</span> <span class='op'>&lt;-</span> <span class='fu'>universalmotif</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/universalmotif/man/create_sequences.html'>create_sequences</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='co'># sequences must have values in their fasta headers</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>seq</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>seq</span><span class='op'>)</span>
<span class='co'># Create random example motif to search for</span>
<span class='va'>motif</span> <span class='op'>&lt;-</span> <span class='fu'>universalmotif</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/universalmotif/man/create_motif.html'>create_motif</a></span><span class='op'>(</span><span class='op'>)</span>

<span class='co'># Search for motif in sequences</span>
<span class='co'># parse_genomic_coord set to FALSE since fasta headers aren't in "chr:start-end" format.</span>
<span class='fu'>runFimo</span><span class='op'>(</span><span class='va'>seq</span>, <span class='va'>motif</span>, parse_genomic_coord <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='co'>#&gt; GRanges object with 2 ranges and 6 metadata columns:
#&gt;       seqnames    ranges strand |    motif_id motif_alt_id     score    pvalue
#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;  &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;
#&gt;   [1]       65     82-91      - |       motif         &lt;NA&gt;   12.1098  4.45e-05
#&gt;   [2]       89     68-77      + |       motif         &lt;NA&gt;   12.4024  3.07e-05
#&gt;          qvalue matched_sequence
#&gt;       &lt;numeric&gt;      &lt;character&gt;
#&gt;   [1]        NA       GTCTATATCT
#&gt;   [2]        NA       GACCATTTCT
#&gt;   -------
#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Spencer Nystrom.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


