<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Motif Scanning using FIMO • memes</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Motif Scanning using FIMO">
<meta property="og:description" content="memes">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">memes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/install_guide.html">Install MEME</a>
    </li>
    <li>
      <a href="../articles/core_dreme.html">De-Novo Motif Discovery</a>
    </li>
    <li>
      <a href="../articles/core_tomtom.html">Motif Comparison</a>
    </li>
    <li>
      <a href="../articles/core_ame.html">Motif Enrichment</a>
    </li>
    <li>
      <a href="../articles/core_fimo.html">Motif Scanning</a>
    </li>
    <li>
      <a href="../articles/tidy_motifs.html">Motif Manipulation</a>
    </li>
    <li>
      <a href="../articles/integrative_analysis.html">ChIP-Seq Data Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/snystrom/memes/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="core_fimo_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Motif Scanning using FIMO</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/snystrom/memes/blob/master/vignettes/core_fimo.Rmd" class="external-link"><code>vignettes/core_fimo.Rmd</code></a></small>
      <div class="hidden name"><code>core_fimo.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/universalmotif/" class="external-link">universalmotif</a></span><span class="op">)</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://snystrom.github.io/memes/" class="external-link">memes</a></span><span class="op">)</span></code></pre></div>
<div id="inputs" class="section level2">
<h2 class="hasAnchor">
<a href="#inputs" class="anchor"></a>Inputs</h2>
<p>FIMO searches input sequences for occurrances of a motif. <code><a href="../reference/runFimo.html">runFimo()</a></code> has two required inputs: fasta-format sequences, with optional genomic coordinate headers, and a set of motifs to detect within the input sequences.</p>
<div id="sequence-inputs" class="section level3">
<h3 class="hasAnchor">
<a href="#sequence-inputs" class="anchor"></a>Sequence Inputs:</h3>
<p>Sequence input to <code><a href="../reference/runFimo.html">runFimo()</a></code> can be as a path to a .fasta formatted file, or as a <code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">Biostrings::XStringSet</a></code> object. Unlike other memes functions, <code><a href="../reference/runFimo.html">runFimo()</a></code> <strong>does not</strong> accept a <code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSetList-class.html" class="external-link">Biostrings::BStringSetList</a></code> as input. This is to simplify ranged join operations (see <a href="#joins">joins</a>) on output data.</p>
<p>By default, <code><a href="../reference/runFimo.html">runFimo()</a></code> will parse genomic coordinates from sequence entries from the fasta headers. These are generated automatically if using <code>get_sequences()</code> to generate sequences for input from a <code>GRanges</code> object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_chip_summits"</span>, package <span class="op">=</span> <span class="st">"memes"</span><span class="op">)</span>

<span class="va">dm.genome</span> <span class="op">&lt;-</span> <span class="fu">BSgenome.Dmelanogaster.UCSC.dm3</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Dmelanogaster.UCSC.dm3/man/package.html" class="external-link">BSgenome.Dmelanogaster.UCSC.dm3</a></span>

<span class="co"># Take 100bp windows around ChIP-seq summits</span>
<span class="va">summit_flank</span> <span class="op">&lt;-</span> <span class="va">example_chip_summits</span> <span class="op">%&gt;%</span> 
  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-anchor.html" class="external-link">anchor_center</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">mutate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> 

<span class="co"># Get sequences in peaks as Biostring::BStringSet</span>
<span class="va">sequences</span> <span class="op">&lt;-</span> <span class="va">summit_flank</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/get_sequence.html">get_sequence</a></span><span class="op">(</span><span class="va">dm.genome</span><span class="op">)</span>

<span class="co"># get_sequence includes genomic coordinate as the fasta header name</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="co">#&gt; [1] "chr3L:40482-40581" "chr3L:49795-49894"</span></code></pre></div>
</div>
<div id="motif-inputs" class="section level3">
<h3 class="hasAnchor">
<a href="#motif-inputs" class="anchor"></a>Motif Inputs:</h3>
<p>Motif input to <code><a href="../reference/runFimo.html">runFimo()</a></code> can be as a path to a .meme formatted file, a list of <code>universalmotif</code> objects, or a singular <code>universalmotif</code> object. <code><a href="../reference/runFimo.html">runFimo()</a></code> will not use any of the default search path behavior for a motif database as in <code><a href="../reference/runAme.html">runAme()</a></code> or <code><a href="../reference/runTomTom.html">runTomTom()</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">e93_motif</span> <span class="op">&lt;-</span> <span class="fu">MotifDb</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/MotifDb/man/MotifDb.html" class="external-link">MotifDb</a></span> <span class="op">%&gt;%</span> 
  <span class="co"># Query the database for the E93 motif using it's gene name</span>
  <span class="fu">MotifDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MotifDb/man/query.html" class="external-link">query</a></span><span class="op">(</span><span class="st">"Eip93F"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># Convert from motifdb format to universalmotif format</span>
  <span class="fu">universalmotif</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/convert_motifs.html" class="external-link">convert_motifs</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># The result is a list, to simplify the object, return it as a single universalmotif</span>
  <span class="va">.</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; See system.file("LICENSE", package="MotifDb") for use restrictions.</span>

<span class="co"># Rename the motif from it's flybase gene number to a more user-friendly name</span>
<span class="va">e93_motif</span><span class="op">[</span><span class="st">"name"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"E93_FlyFactor"</span></code></pre></div>
</div>
</div>
<div id="note-about-default-settings" class="section level2">
<h2 class="hasAnchor">
<a href="#note-about-default-settings" class="anchor"></a>Note about default settings</h2>
<p><code><a href="../reference/runFimo.html">runFimo()</a></code> is configured to use different default behavior relative to the commandline and MEME-Suite Server versions. By default, <code><a href="../reference/runFimo.html">runFimo()</a></code> runs using <code>text</code> mode, which greatly increases speed and allows returning all detected matches to the input motifs. By default <code><a href="../reference/runFimo.html">runFimo()</a></code> will add the sequence of the matched region to the output data; however, this operation can be very slow for large sets of regions and can drastically increase the size of the output data. To speed up the operation and decrease data size, set <code>skip_matched_sequence = TRUE</code>. Sequence information can be added back later using <code><a href="../reference/add_sequence.html">add_sequence()</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fimo_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runFimo.html">runFimo</a></span><span class="op">(</span><span class="va">sequences</span>, <span class="va">e93_motif</span><span class="op">)</span></code></pre></div>
</div>
<div id="joins" class="section level2">
<h2 class="hasAnchor">
<a href="#joins" class="anchor"></a>Data integration with join operations</h2>
<p>The <code>plyranges</code> package provides an extended framework for performing range-based operations in R. While several of its utilities are useful for range-based analyses, the <code>join_</code> functions are particularly useful for integrating FIMO results with input peak information. A few common examples are briefly highlighted below:</p>
<p><code><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">plyranges::join_overlap_left()</a></code> can be used to add peak-level metadata to motif position information:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fimo_results_with_peak_info</span> <span class="op">&lt;-</span> <span class="va">fimo_results</span> <span class="op">%&gt;%</span> 
  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_left</a></span><span class="op">(</span><span class="va">summit_flank</span><span class="op">)</span>

<span class="va">fimo_results_with_peak_info</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt; GRanges object with 5 ranges and 9 metadata columns:</span>
<span class="co">#&gt;       seqnames        ranges strand |      motif_id           motif_alt_id</span>
<span class="co">#&gt;          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;character&gt;            &lt;character&gt;</span>
<span class="co">#&gt;   [1]    chr3L   66310-66321      - | E93_FlyFactor Eip93F_SANGER_10_FBg..</span>
<span class="co">#&gt;   [2]    chr3L 138527-138538      - | E93_FlyFactor Eip93F_SANGER_10_FBg..</span>
<span class="co">#&gt;   [3]    chr3L 384344-384355      - | E93_FlyFactor Eip93F_SANGER_10_FBg..</span>
<span class="co">#&gt;   [4]    chr3L 714116-714127      - | E93_FlyFactor Eip93F_SANGER_10_FBg..</span>
<span class="co">#&gt;   [5]    chr3L 714145-714156      - | E93_FlyFactor Eip93F_SANGER_10_FBg..</span>
<span class="co">#&gt;           score    pvalue    qvalue matched_sequence          id</span>
<span class="co">#&gt;       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;      &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt;   [1]     14.25  1.01e-05        NA     CCGGCCGAAAAT   peak_4627</span>
<span class="co">#&gt;   [2]     10.10  8.96e-05        NA     CTGTCCAAAAAT   peak_4647</span>
<span class="co">#&gt;   [3]     10.87  6.65e-05        NA     GCCGCCGAAAAA   peak_4670</span>
<span class="co">#&gt;   [4]     12.90  2.43e-05        NA     CGTGCCAAAAAT   peak_4752</span>
<span class="co">#&gt;   [5]     10.25  8.51e-05        NA     CTTCCCAAAAAA   peak_4752</span>
<span class="co">#&gt;       peak_binding_description e93_sensitive_behavior</span>
<span class="co">#&gt;                    &lt;character&gt;            &lt;character&gt;</span>
<span class="co">#&gt;   [1]                  ectopic                 Static</span>
<span class="co">#&gt;   [2]                   orphan                 Static</span>
<span class="co">#&gt;   [3]                   orphan             Decreasing</span>
<span class="co">#&gt;   [4]                  entopic                 Static</span>
<span class="co">#&gt;   [5]                  entopic                 Static</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></code></pre></div>
<p><code>plyranges::intersect_()</code> can be used to simultaneously subset input peaks to the ranges overlapping motif hits while appending motif-level metadata to each overlap.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">input_intersect_hits</span> <span class="op">&lt;-</span> <span class="va">summit_flank</span> <span class="op">%&gt;%</span> 
  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_intersect</a></span><span class="op">(</span><span class="va">fimo_results</span><span class="op">)</span>

<span class="va">input_intersect_hits</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt; GRanges object with 5 ranges and 9 metadata columns:</span>
<span class="co">#&gt;       seqnames        ranges strand |          id peak_binding_description</span>
<span class="co">#&gt;          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;              &lt;character&gt;</span>
<span class="co">#&gt;   [1]    chr3L   66310-66321      * |   peak_4627                  ectopic</span>
<span class="co">#&gt;   [2]    chr3L 138527-138538      * |   peak_4647                   orphan</span>
<span class="co">#&gt;   [3]    chr3L 384344-384355      * |   peak_4670                   orphan</span>
<span class="co">#&gt;   [4]    chr3L 714116-714127      * |   peak_4752                  entopic</span>
<span class="co">#&gt;   [5]    chr3L 714145-714156      * |   peak_4752                  entopic</span>
<span class="co">#&gt;       e93_sensitive_behavior      motif_id           motif_alt_id     score</span>
<span class="co">#&gt;                  &lt;character&gt;   &lt;character&gt;            &lt;character&gt; &lt;numeric&gt;</span>
<span class="co">#&gt;   [1]                 Static E93_FlyFactor Eip93F_SANGER_10_FBg..     14.25</span>
<span class="co">#&gt;   [2]                 Static E93_FlyFactor Eip93F_SANGER_10_FBg..     10.10</span>
<span class="co">#&gt;   [3]             Decreasing E93_FlyFactor Eip93F_SANGER_10_FBg..     10.87</span>
<span class="co">#&gt;   [4]                 Static E93_FlyFactor Eip93F_SANGER_10_FBg..     12.90</span>
<span class="co">#&gt;   [5]                 Static E93_FlyFactor Eip93F_SANGER_10_FBg..     10.25</span>
<span class="co">#&gt;          pvalue    qvalue matched_sequence</span>
<span class="co">#&gt;       &lt;numeric&gt; &lt;numeric&gt;      &lt;character&gt;</span>
<span class="co">#&gt;   [1]  1.01e-05        NA     CCGGCCGAAAAT</span>
<span class="co">#&gt;   [2]  8.96e-05        NA     CTGTCCAAAAAT</span>
<span class="co">#&gt;   [3]  6.65e-05        NA     GCCGCCGAAAAA</span>
<span class="co">#&gt;   [4]  2.43e-05        NA     CGTGCCAAAAAT</span>
<span class="co">#&gt;   [5]  8.51e-05        NA     CTTCCCAAAAAA</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></code></pre></div>
</div>
<div id="matched-sequence" class="section level2">
<h2 class="hasAnchor">
<a href="#matched-sequence" class="anchor"></a>Identifying matched sequence</h2>
<p>When setting <code>skip_match_sequence = TRUE</code>, FIMO does not automatically return the matched sequence within each hit. These sequences can be easily recovered in R using <code><a href="../reference/add_sequence.html">add_sequence()</a></code> on the FIMO results <code>GRanges</code> object.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fimo_results_with_seq</span> <span class="op">&lt;-</span> <span class="va">fimo_results</span> <span class="op">%&gt;%</span> 
  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_left</a></span><span class="op">(</span><span class="va">summit_flank</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/add_sequence.html">add_sequence</a></span><span class="op">(</span><span class="va">dm.genome</span><span class="op">)</span></code></pre></div>
<p>Returning the sequence of the matched regions can be used to re-derive PWMs from different match categories as follows (here done for different binding categories):</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">motifs_by_binding</span> <span class="op">&lt;-</span> <span class="va">fimo_results_with_seq</span> <span class="op">%&gt;%</span> 
  <span class="co"># Split on parameter of interest</span>
  <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">$</span><span class="va">peak_binding_description</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># Convert GRangesList to regular list() to use `purrr`</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># imap passes the list entry as .x and the name of that object to .y</span>
  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/imap.html" class="external-link">imap</a></span><span class="op">(</span><span class="op">~</span><span class="op">{</span>
    <span class="co"># Pass the sequence column to create_motif to generate a PCM</span>
    <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/create_motif.html" class="external-link">create_motif</a></span><span class="op">(</span><span class="va">.x</span><span class="op">$</span><span class="va">sequence</span>, 
                 <span class="co"># Append the binding description to the motif name</span>
                 name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"E93_"</span>, <span class="va">.y</span><span class="op">)</span><span class="op">)</span>
    <span class="op">}</span><span class="op">)</span></code></pre></div>
<p>Motifs from each category can be visualized with <code><a href="https://rdrr.io/pkg/universalmotif/man/view_motifs.html" class="external-link">universalmotif::view_motifs()</a></code></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">motifs_by_binding</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/view_motifs.html" class="external-link">view_motifs</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="core_fimo_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>To allow better comparison to the reference motif, we can append it to the list as follows:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">motifs_by_binding</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>
  <span class="co"># Add the E93 FlyFactor motif to the list as a reference</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"E93_FlyFactor"</span> <span class="op">=</span> <span class="va">e93_motif</span><span class="op">)</span>,
  <span class="va">motifs_by_binding</span>
<span class="op">)</span></code></pre></div>
<p>Visualizing the motifs as ICMs reveals subtle differences in E93 motif sequence between each category.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">motifs_by_binding</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/view_motifs.html" class="external-link">view_motifs</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="core_fimo_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>Visualizing the results as a position-probability matrix (PPM) does a better job of demonstrating that the primary differences between each category are coming from positions 1-4 in the matched sequences.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">motifs_by_binding</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/view_motifs.html" class="external-link">view_motifs</a></span><span class="op">(</span>use.type <span class="op">=</span> <span class="st">"PPM"</span><span class="op">)</span></code></pre></div>
<p><img src="core_fimo_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>Finally, the sequence-level information can be used to visualize all sequences and their contribution to the final PWM using <code>plot_sequence_heatmap</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_sequence_heatmap.html">plot_sequence_heatmap</a></span><span class="op">(</span><span class="va">fimo_results_with_seq</span><span class="op">$</span><span class="va">sequence</span><span class="op">)</span></code></pre></div>
<p><img src="core_fimo_files/figure-html/unnamed-chunk-13-1.png" width="288"></p>
</div>
<div id="importing-data-from-previous-fimo-runs" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-data-from-previous-fimo-runs" class="anchor"></a>Importing Data from previous FIMO Runs</h2>
<p><code><a href="../reference/importFimo.html">importFimo()</a></code> can be used to import an <code>fimo.tsv</code> file from a previous run on the MEME server or on the commandline. Details for how to save data from the FIMO webserver are below.</p>
<div id="saving-data-from-fimo-web-server" class="section level3">
<h3 class="hasAnchor">
<a href="#saving-data-from-fimo-web-server" class="anchor"></a>Saving data from FIMO Web Server</h3>
<p>To download TSV data from the FIMO Server, right-click the FIMO TSV output link and “Save Target As” or “Save Link As” (see example image below), and save as <code>&lt;filename&gt;.tsv</code>. This file can be read using <code><a href="../reference/importFimo.html">importFimo()</a></code>.</p>
<p><img src="save_fimo.png"></p>
</div>
</div>
<div id="citation" class="section level1">
<h1 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h1>
<p>memes is a wrapper for a select few tools from the MEME Suite, which were developed by another group. In addition to citing memes, please cite the MEME Suite tools corresponding to the tools you use.</p>
<p>If you use <code><a href="../reference/runFimo.html">runFimo()</a></code> in your analysis, please cite:</p>
<p>Charles E. Grant, Timothy L. Bailey, and William Stafford Noble, “FIMO: Scanning for occurrences of a given motif”, Bioinformatics, 27(7):1017-1018, 2011. <a href="http://bioinformatics.oxfordjournals.org/content/early/2011/02/16/bioinformatics.btr064.full" class="external-link">full text</a></p>
<div id="licensing-restrictions" class="section level2">
<h2 class="hasAnchor">
<a href="#licensing-restrictions" class="anchor"></a>Licensing Restrictions</h2>
<p>The MEME Suite is free for non-profit use, but for-profit users should purchase a license. See the <a href="http://meme-suite.org/doc/copyright.html" class="external-link">MEME Suite Copyright Page</a> for details.</p>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R Under development (unstable) (2021-03-18 r80099)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.2 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] memes_0.99.5          universalmotif_1.9.39 magrittr_2.0.1       </span>
<span class="co">#&gt; [4] GenomicRanges_1.43.3  GenomeInfoDb_1.27.8   IRanges_2.25.6       </span>
<span class="co">#&gt; [7] S4Vectors_0.29.11     BiocGenerics_0.37.1  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] bitops_1.0-6                          matrixStats_0.58.0                   </span>
<span class="co">#&gt;  [3] fs_1.5.0                              rprojroot_2.0.2                      </span>
<span class="co">#&gt;  [5] tools_4.1.0                           bslib_0.2.4                          </span>
<span class="co">#&gt;  [7] utf8_1.2.1                            R6_2.5.0                             </span>
<span class="co">#&gt;  [9] splitstackshape_1.4.8                 colorspace_2.0-0                     </span>
<span class="co">#&gt; [11] withr_2.4.1                           processx_3.5.0                       </span>
<span class="co">#&gt; [13] tidyselect_1.1.0                      compiler_4.1.0                       </span>
<span class="co">#&gt; [15] textshaping_0.3.3                     Biobase_2.51.0                       </span>
<span class="co">#&gt; [17] desc_1.3.0                            DelayedArray_0.17.10                 </span>
<span class="co">#&gt; [19] labeling_0.4.2                        rtracklayer_1.51.5                   </span>
<span class="co">#&gt; [21] sass_0.3.1                            scales_1.1.1                         </span>
<span class="co">#&gt; [23] readr_1.4.0                           pkgdown_1.6.1.9001                   </span>
<span class="co">#&gt; [25] systemfonts_1.0.1                     stringr_1.4.0                        </span>
<span class="co">#&gt; [27] digest_0.6.27                         Rsamtools_2.7.1                      </span>
<span class="co">#&gt; [29] rmarkdown_2.7                         R.utils_2.10.1                       </span>
<span class="co">#&gt; [31] XVector_0.31.1                        pkgconfig_2.0.3                      </span>
<span class="co">#&gt; [33] htmltools_0.5.1.1                     MatrixGenerics_1.3.1                 </span>
<span class="co">#&gt; [35] highr_0.8                             fastmap_1.1.0                        </span>
<span class="co">#&gt; [37] BSgenome_1.59.2                       rlang_0.4.10                         </span>
<span class="co">#&gt; [39] farver_2.1.0                          BSgenome.Dmelanogaster.UCSC.dm3_1.4.0</span>
<span class="co">#&gt; [41] jquerylib_0.1.3                       BiocIO_1.1.2                         </span>
<span class="co">#&gt; [43] generics_0.1.0                        jsonlite_1.7.2                       </span>
<span class="co">#&gt; [45] BiocParallel_1.25.5                   dplyr_1.0.5                          </span>
<span class="co">#&gt; [47] R.oo_1.24.0                           RCurl_1.98-1.3                       </span>
<span class="co">#&gt; [49] GenomeInfoDbData_1.2.4                patchwork_1.1.1                      </span>
<span class="co">#&gt; [51] Matrix_1.3-2                          Rcpp_1.0.6                           </span>
<span class="co">#&gt; [53] waldo_0.2.5                           munsell_0.5.0                        </span>
<span class="co">#&gt; [55] fansi_0.4.2                           lifecycle_1.0.0                      </span>
<span class="co">#&gt; [57] R.methodsS3_1.8.1                     stringi_1.5.3                        </span>
<span class="co">#&gt; [59] yaml_2.2.1                            MASS_7.3-53.1                        </span>
<span class="co">#&gt; [61] SummarizedExperiment_1.21.1           zlibbioc_1.37.0                      </span>
<span class="co">#&gt; [63] grid_4.1.0                            ggseqlogo_0.1                        </span>
<span class="co">#&gt; [65] crayon_1.4.1                          lattice_0.20-41                      </span>
<span class="co">#&gt; [67] Biostrings_2.59.2                     hms_1.0.0                            </span>
<span class="co">#&gt; [69] ps_1.6.0                              knitr_1.31                           </span>
<span class="co">#&gt; [71] pillar_1.5.1                          rjson_0.2.20                         </span>
<span class="co">#&gt; [73] pkgload_1.2.0                         XML_3.99-0.6                         </span>
<span class="co">#&gt; [75] glue_1.4.2                            evaluate_0.14                        </span>
<span class="co">#&gt; [77] data.table_1.14.0                     vctrs_0.3.7                          </span>
<span class="co">#&gt; [79] testthat_3.0.2                        gtable_0.3.0                         </span>
<span class="co">#&gt; [81] purrr_0.3.4                           tidyr_1.1.3                          </span>
<span class="co">#&gt; [83] rematch2_2.1.2                        cachem_1.0.4                         </span>
<span class="co">#&gt; [85] ggplot2_3.3.3                         xfun_0.22                            </span>
<span class="co">#&gt; [87] restfulr_0.0.13                       ragg_1.1.2                           </span>
<span class="co">#&gt; [89] MotifDb_1.33.0                        tibble_3.1.0                         </span>
<span class="co">#&gt; [91] diffobj_0.3.4                         GenomicAlignments_1.27.2             </span>
<span class="co">#&gt; [93] plyranges_1.11.0                      memoise_2.0.0                        </span>
<span class="co">#&gt; [95] cmdfun_1.0.2                          ellipsis_0.3.1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Spencer Nystrom.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
