<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Motif Enrichment Testing using AME • memes</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Motif Enrichment Testing using AME">
<meta property="og:description" content="memes">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">memes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/install_guide.html">Install MEME</a>
    </li>
    <li>
      <a href="../articles/core_dreme.html">De-Novo Motif Discovery</a>
    </li>
    <li>
      <a href="../articles/core_tomtom.html">Motif Comparison</a>
    </li>
    <li>
      <a href="../articles/core_ame.html">Motif Enrichment</a>
    </li>
    <li>
      <a href="../articles/core_fimo.html">Motif Scanning</a>
    </li>
    <li>
      <a href="../articles/tidy_motifs.html">Motif Manipulation</a>
    </li>
    <li>
      <a href="../articles/integrative_analysis.html">ChIP-Seq Data Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/snystrom/memes/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="core_ame_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Motif Enrichment Testing using AME</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/snystrom/memes/blob/master/vignettes/core_ame.Rmd" class="external-link"><code>vignettes/core_ame.Rmd</code></a></small>
      <div class="hidden name"><code>core_ame.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://snystrom.github.io/memes/" class="external-link">memes</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></code></pre></div>
<div id="sequence-input" class="section level2">
<h2 class="hasAnchor">
<a href="#sequence-input" class="anchor"></a>Sequence Input</h2>
<p>AME requires a series of input sequences to scan for motif enrichment. <code><a href="../reference/runAme.html">runAme()</a></code> accepts sequence input in the following formats:</p>
<ul>
<li>a <code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">Biostrings::XStringSet</a></code> object</li>
<li>a named list of <code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">Biostrings::XStringSet</a></code> objects</li>
<li>a path to a .fasta file</li>
</ul>
<p><strong>NOTE</strong> <code>XStringSet</code> inputs can be easily generated for DNA sequences from a GRanges object using the <code><a href="../reference/get_sequence.html">get_sequence()</a></code> function</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_peaks"</span>, package <span class="op">=</span> <span class="st">"memes"</span><span class="op">)</span>

<span class="va">dm.genome</span> <span class="op">&lt;-</span> <span class="fu">BSgenome.Dmelanogaster.UCSC.dm3</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Dmelanogaster.UCSC.dm3/man/package.html" class="external-link">BSgenome.Dmelanogaster.UCSC.dm3</a></span>

<span class="va">sequence</span> <span class="op">&lt;-</span> <span class="va">example_peaks</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/get_sequence.html">get_sequence</a></span><span class="op">(</span><span class="va">dm.genome</span><span class="op">)</span></code></pre></div>
</div>
<div id="database-input" class="section level2">
<h2 class="hasAnchor">
<a href="#database-input" class="anchor"></a>Database Input</h2>
<p>AME scans input sequences against a database of known motifs and tests for enrichment of each motif in the database. <code><a href="../reference/runAme.html">runAme()</a></code> can accept a database in the following formats:</p>
<ul>
<li>a list of universalmotif objects</li>
<li>a single universalmotif object</li>
<li>the results object from <code>runDreme</code>
</li>
<li>a path to a .meme format file</li>
</ul>
<div id="setting-a-default-database" class="section level3">
<h3 class="hasAnchor">
<a href="#setting-a-default-database" class="anchor"></a>Setting a default database</h3>
<p>memes can be configured to use a default .meme format file as the query database, which it will use if the user does not provide a value to <code>database</code> when calling <code><a href="../reference/runAme.html">runAme()</a></code>. The following locations will be searched in order.</p>
<ol style="list-style-type: decimal">
<li>The <code>meme_db</code> option, defined using <code><a href="https://rdrr.io/r/base/options.html" class="external-link">options(meme_db = "path/to/database.meme")</a></code>
</li>
</ol>
<ul>
<li>The <code>meme_db</code> option can also be set to an R object, like a universalmotif list.</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>The <code>MEME_DB</code> environment variable defined in <code>.Renviron</code>
</li>
</ol>
<ul>
<li>The <code>MEME_DB</code> variable will only accept a path to a .meme file</li>
</ul>
<p><strong>NOTE:</strong> if an invalid location is found at one option, <code><a href="../reference/runAme.html">runAme()</a></code> will fall back to the next location if valid (eg if the <code>meme_db</code> option is set to an invalid file, but the <code>MEME_DB</code> environment variable is a valid file, the <code>MEME_DB</code> path will be used.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>meme_db <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/flyFactorSurvey_cleaned.meme"</span>, package <span class="op">=</span> <span class="st">"memes"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="running-ame" class="section level2">
<h2 class="hasAnchor">
<a href="#running-ame" class="anchor"></a>Running AME</h2>
<p><code><a href="../reference/runAme.html">runAme()</a></code> supports running AME using three modes:</p>
<table class="table">
<colgroup>
<col width="19%">
<col width="29%">
<col width="51%">
</colgroup>
<thead><tr class="header">
<th align="center">AME Mode</th>
<th align="center">Description</th>
<th align="left">Command</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">Vs Shuffled</td>
<td align="center">Input vs Shuffled Sequence</td>
<td align="left"><code><a href="../reference/runAme.html">runAme(input = sequence, control = "shuffle")</a></code></td>
</tr>
<tr class="even">
<td align="center">Discriminative</td>
<td align="center">Input vs Control Sequence</td>
<td align="left"><code><a href="../reference/runAme.html">runAme(input = sequence, control = control)</a></code></td>
</tr>
<tr class="odd">
<td align="center">Partitioning</td>
<td align="center">Rank Input by fasta score</td>
<td align="left"><code><a href="../reference/runAme.html">runAme(input = sequence, control = NA)</a></code></td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ame_vs_shuffle</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runAme.html">runAme</a></span><span class="op">(</span><span class="va">sequence</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ame_vs_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runAme.html">runAme</a></span><span class="op">(</span><span class="va">sequence</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, <span class="va">sequence</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p>To run AME using partitioning mode, the fasta header must contain a score value for each entry in the form: “&gt;entry_name score”. The <code>get_sequences()</code> <code>score</code> argument allows users to set the score value to a column value from input regions.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sequence_scored</span> <span class="op">&lt;-</span> <span class="va">example_peaks</span> <span class="op">%&gt;%</span> 
  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">mutate</a></span><span class="op">(</span>score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/get_sequence.html">get_sequence</a></span><span class="op">(</span><span class="va">dm.genome</span>, score <span class="op">=</span> <span class="st">"score"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sequence_scored</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>
<span class="co">#&gt; [1] "chr3L:14551117-14551323 1"</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ame_partition</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runAme.html">runAme</a></span><span class="op">(</span><span class="va">sequence_scored</span>, control <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></code></pre></div>
<div id="running-ame-on-multiple-groups" class="section level3">
<h3 class="hasAnchor">
<a href="#running-ame-on-multiple-groups" class="anchor"></a>Running AME on multiple groups</h3>
<p>If using a list input to <code><a href="../reference/runAme.html">runAme()</a></code>, it will dispatch multiple AME runs for each object in the list.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_chip_summits"</span>, package <span class="op">=</span> <span class="st">"memes"</span><span class="op">)</span>

<span class="va">seq_by_behavior</span> <span class="op">&lt;-</span> <span class="va">example_chip_summits</span> <span class="op">%&gt;%</span> 
  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">mutate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">$</span><span class="va">e93_sensitive_behavior</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/get_sequence.html">get_sequence</a></span><span class="op">(</span><span class="va">dm.genome</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ame_by_behavior</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runAme.html">runAme</a></span><span class="op">(</span><span class="va">seq_by_behavior</span><span class="op">)</span></code></pre></div>
<div id="discriminative-analysis-using-list-input" class="section level4">
<h4 class="hasAnchor">
<a href="#discriminative-analysis-using-list-input" class="anchor"></a>Discriminative analysis using list input</h4>
<p>If the input to <code><a href="../reference/runAme.html">runAme()</a></code> is a named list of <code>XStringSet</code> objects, <code>control</code> can be set to one or more values from <code><a href="https://rdrr.io/r/base/names.html" class="external-link">names(input)</a></code> to use those regions as background. It will skip running those regions as the input. The following code will result in these comparisons:</p>
<ol style="list-style-type: decimal">
<li>Increasing vs Static</li>
<li>Decreasing vs Static</li>
</ol>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ame_by_behavior_vs_static</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runAme.html">runAme</a></span><span class="op">(</span><span class="va">seq_by_behavior</span>, control <span class="op">=</span> <span class="st">"Static"</span><span class="op">)</span></code></pre></div>
<p>If multiple names are used in the <code>control</code> section, they will be combined together to make a single control set which will be used for all comparisons. Here, we use “Static” and “Decreasing” sites as the control, which will result in only running 1 comparison: Increasing vs Static+Decreasing.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/runAme.html">runAme</a></span><span class="op">(</span><span class="va">seq_by_behavior</span>, control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Static"</span>, <span class="st">"Decreasing"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
</div>
<div id="output-format" class="section level2">
<h2 class="hasAnchor">
<a href="#output-format" class="anchor"></a>Output Format</h2>
<p>AME will return different output formats depending on the <code>method</code> used. For detailed information about these values see the <a href="http://meme-suite.org/doc/ame-output-format.html" class="external-link">AME Output description webpage</a>. As a general rule of thumb, <code><a href="../reference/runAme.html">runAme()</a></code> will return the same column names described in the webpage, except dashes are removed and all column names are lowercase.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ame_by_behavior_vs_static</span><span class="op">$</span><span class="va">Decreasing</span> <span class="op">%&gt;%</span> 
  <span class="va">names</span>
<span class="co">#&gt;  [1] "rank"         "motif_db"     "motif_id"     "motif_alt_id" "consensus"   </span>
<span class="co">#&gt;  [6] "pvalue"       "adj.pvalue"   "evalue"       "tests"        "fasta_max"   </span>
<span class="co">#&gt; [11] "pos"          "neg"          "pwm_min"      "tp"           "tp_percent"  </span>
<span class="co">#&gt; [16] "fp"           "fp_percent"</span></code></pre></div>
<p>If <code><a href="../reference/runAme.html">runAme()</a></code> is run with <code>method = "fisher"</code>, the sequences output can be added to the results by setting <code>sequences = TRUE</code>. This will be added as a list column named <code>sequences</code> that can be unnested using <code><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">tidyr::unnest()</a></code>.</p>
</div>
<div id="visualizing-results-as-heatmap" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-results-as-heatmap" class="anchor"></a>Visualizing Results as Heatmap</h2>
<p>The <code><a href="../reference/plot_ame_heatmap.html">plot_ame_heatmap()</a></code> function provides a method to easily generate visualizations of AME results.</p>
<p>To plot results from multiple runs together, they must first be joined into 1 data frame. The <code>ame_by_behavior_vs_static</code> object is a list whose names correspond to the E93 response (Increasing or Decreasing). The list can be combined into a data.frame using <code><a href="https://dplyr.tidyverse.org/reference/bind.html" class="external-link">dplyr::bind_rows</a></code>. Setting <code>.id = "behavior</code> creates a new column <code>behavior</code> that contains the names from the <code>ame_by_behavior_vs_static</code> list. In this way, the resulting data.frame contains all AME results for each run, which can be distinguished by the <code>behavior</code> column.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ame_by_behavior_vs_static</span> <span class="op">%&gt;%</span> 
  <span class="co"># AME results in list format are easily combined using dplyr::bind_rows</span>
  <span class="co"># .id will specify a column to hold the list object names</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html" class="external-link">bind_rows</a></span><span class="op">(</span>.id <span class="op">=</span> <span class="st">"behavior"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># setting group to a column name will split the results on the y-axis</span>
  <span class="fu"><a href="../reference/plot_ame_heatmap.html">plot_ame_heatmap</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">behavior</span><span class="op">)</span></code></pre></div>
<p><img src="core_ame_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<div id="complex-heatmap-example" class="section level3">
<h3 class="hasAnchor">
<a href="#complex-heatmap-example" class="anchor"></a>Complex Heatmap Example</h3>
<p>There are several nuances when making heatmap visualizations of these data. The following examples highlight some of these issues and provide alternative approaches and solutions.</p>
<p>We start by using different binding site categories as input.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seq_by_binding</span> <span class="op">&lt;-</span> <span class="va">example_chip_summits</span> <span class="op">%&gt;%</span> 
  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">mutate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">$</span><span class="va">peak_binding_description</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/get_sequence.html">get_sequence</a></span><span class="op">(</span><span class="va">dm.genome</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ame_by_binding</span> <span class="op">&lt;-</span> <span class="va">seq_by_binding</span> <span class="op">%&gt;%</span> 
  <span class="va">runAme</span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ame_res</span> <span class="op">&lt;-</span> <span class="va">ame_by_binding</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html" class="external-link">bind_rows</a></span><span class="op">(</span>.id <span class="op">=</span> <span class="st">"binding_type"</span><span class="op">)</span></code></pre></div>
<p>It is possible to aggregate results from multiple runs into a heatmap by setting the <code>group</code> parameter in <code><a href="../reference/plot_ame_heatmap.html">plot_ame_heatmap()</a></code>.</p>
<p>This is too many hits to properly view in this vignette, but you can see that the heatmap will plot motifs by their overlap across groups, where unique motifs are on the left, and shared motifs are on the right.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ame_res</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/plot_ame_heatmap.html">plot_ame_heatmap</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">binding_type</span><span class="op">)</span></code></pre></div>
<p><img src="core_ame_files/figure-html/unnamed-chunk-19-1.png" width="1440"></p>
</div>
</div>
<div id="issues-with-heatmap-visualization" class="section level2">
<h2 class="hasAnchor">
<a href="#issues-with-heatmap-visualization" class="anchor"></a>Issues with Heatmap Visualization</h2>
<p>The dynamic range of p-values in these data varies between groups. For this reason, a simple heatmap scaled using all data values will make it more difficult to interpret within groups with a lower dynamic range of values. In other words, because the dynamic range of values are different between experiments, placing them on the default scale for comparison may not always be the most optimal visualization.</p>
<p>We can partially overcome this limitation by filling the heatmap with the normalized rank value for each TF, which accounts for differences in total number of discovered motifs between AME runs. Although it does not completely abrogate differences, the signal values for high-ranked motifs within groups will be more comparable. However, <strong>the normalized rank visualization eliminates all real values related to statistical significance!</strong> Instead, this visualization represents the relative ranks of hits within an AME run, which already pass a significance threshold set during <code>runAME()</code>. This means that even if several motifs have similar or even identical p-values, their heatmap representation will be a different color value based on their ranked order in the results list. This tends to only be useful when there are a large number of hits (&gt;=100). Both visualizations can be useful and reveal different properties of the data to the user. <strong>If in doubt</strong>, prefer the <code>-log10(adj.pvalue)</code> representation.</p>
<p>Below is a comparison of the distribution of values when using <code>-log10(adj.pvalue)</code> (A) vs normalized ranks (B). Because orphan sites tend to have smaller p-values overall, the heatmap scale will be skewed towards the high values in the orphan data, making ectopic and entopic heat values lighter by comparison.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ame_res</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ame_compare_heatmap_methods.html">ame_compare_heatmap_methods</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">binding_type</span><span class="op">)</span></code></pre></div>
<p><img src="core_ame_files/figure-html/unnamed-chunk-20-1.png" width="720"></p>
<p>To use the normalized rank value, set <code>value = "normalize"</code> in <code><a href="../reference/plot_ame_heatmap.html">plot_ame_heatmap()</a></code>.</p>
<p>This plot reveals that the motifs which tend to be shared across all 3 categories tend to be higher ranks in the output than the motifs unique to the different categories, which tend to come from lower ranks. This <em>suggests</em> that although there are differences in motif content across the three categories, they may be largely similar in motif makeup. We will investigate this question in more detail in the “Denovo motif similarity” section.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>

<span class="op">(</span><span class="va">normalize_heatmap</span> <span class="op">&lt;-</span> <span class="va">ame_res</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">binding_type</span>, <span class="va">motif_alt_id</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">adj.pvalue</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">adj.pvalue</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/plot_ame_heatmap.html">plot_ame_heatmap</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">binding_type</span>, id <span class="op">=</span> <span class="va">motif_alt_id</span>, value <span class="op">=</span> <span class="st">"normalize"</span><span class="op">)</span> <span class="op">+</span>
    <span class="co"># All ggplot functions can be used to extend or edit the heatmap plots</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"value = \"normalize\""</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="core_ame_files/figure-html/unnamed-chunk-21-1.png" width="2400"></p>
<p>An additional third option exists to rescale the <code>-log10(adj.pvalue)</code> heatmap to change the heatmap’s maxiumum color value. This allows the user to maintain values which represent significance, but rescale the data to capture the lower end of the dynamic range. Using the cumulative distribution plot above, a reasonable cutoff is anywhere between 7 &amp; 10, which captures &gt; 90% of the data for ectopic and entopic sites.</p>
<p>A comparison of all three methods can be seen below.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pval_heatmap</span> <span class="op">&lt;-</span> <span class="va">ame_res</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">binding_type</span>, <span class="va">motif_alt_id</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">adj.pvalue</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">adj.pvalue</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/plot_ame_heatmap.html">plot_ame_heatmap</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">binding_type</span>, id <span class="op">=</span> <span class="va">motif_alt_id</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"value = -log10(adj.pvalue)"</span><span class="op">)</span>

<span class="va">scale_heatmap</span> <span class="op">&lt;-</span> <span class="va">ame_res</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">binding_type</span>, <span class="va">motif_alt_id</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">adj.pvalue</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">adj.pvalue</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/plot_ame_heatmap.html">plot_ame_heatmap</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">binding_type</span>, id <span class="op">=</span> <span class="va">motif_alt_id</span>, scale_max <span class="op">=</span> <span class="fl">7.5</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"value = -log10(adj.pvalue) (scale capped at 7.5)"</span><span class="op">)</span></code></pre></div>
<p>Below is a comparison using the <code>-log10(adj.pvalue)</code> vs <code>normalize</code> methods for plotting the heatmap. Note how the different plots highlight different data properties. The <code>-log10(adj.pvalue)</code> plot shows overall significance of each hit, while <code>normalize</code> method shows the relative rank of each hit within a <code>binding_type</code>. Lowering the maximum scale value in C) does a better job than A) at visualizing differences in significance along the ectopic and entopic rows at the cost of decreasing the dynamic range of the orphan row. Selecting a visualization for publication will depend heavily on context, but if in doubt, prefer one which includes information of statistical significance as in A) or C).</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">pval_heatmap</span>, 
                   <span class="va">normalize_heatmap</span>, 
                   <span class="va">scale_heatmap</span>, 
                   ncol <span class="op">=</span> <span class="fl">1</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></code></pre></div>
<p><img src="core_ame_files/figure-html/unnamed-chunk-23-1.png" width="2400"></p>
</div>
<div id="importing-previous-data" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-previous-data" class="anchor"></a>Importing Previous Data</h2>
<p><code><a href="../reference/importAme.html">importAme()</a></code> can be used to import an <code>ame.tsv</code> file from a previous run on the MEME server or on the commandline. Details for how to save data from the AME webserver are below.</p>
<p>Optionally, if AME was run on the commandline with <code>--method fisher</code>, the user can pass a path to the <code>sequences.tsv</code> file to the <code>sequences</code> argument of <code><a href="../reference/importAme.html">importAme()</a></code> to append the sequence information to the AME results.</p>
<div id="saving-data-from-ame-web-server" class="section level3">
<h3 class="hasAnchor">
<a href="#saving-data-from-ame-web-server" class="anchor"></a>Saving data from AME Web Server</h3>
<p>To download TSV data from the MEME Server, right-click the AME TSV output link and “Save Target As” or “Save Link As” (see example image below), and save as <code>&lt;filename&gt;.tsv</code>. This file can be read using <code><a href="../reference/importAme.html">importAme()</a></code>.</p>
<div class="figure">
<img src="save_ame.png" alt=""><p class="caption">saving AME tsv results</p>
</div>
</div>
</div>
<div id="citation" class="section level1">
<h1 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h1>
<p>memes is a wrapper for a select few tools from the MEME Suite, which were developed by another group. In addition to citing memes, please cite the MEME Suite tools corresponding to the tools you use.</p>
<p>If you use <code><a href="../reference/runAme.html">runAme()</a></code> in your analysis, please cite:</p>
<p>Robert McLeay and Timothy L. Bailey, “Motif Enrichment Analysis: A unified framework and method evaluation”, BMC Bioinformatics, 11:165, 2010, <a href="doi:10.1186/1471-2105-11-165" class="uri">doi:10.1186/1471-2105-11-165</a>. <a href="http://www.biomedcentral.com/1471-2105/11/165" class="external-link">full text</a></p>
<div id="licensing-restrictions" class="section level2">
<h2 class="hasAnchor">
<a href="#licensing-restrictions" class="anchor"></a>Licensing Restrictions</h2>
<p>The MEME Suite is free for non-profit use, but for-profit users should purchase a license. See the <a href="http://meme-suite.org/doc/copyright.html" class="external-link">MEME Suite Copyright Page</a> for details.</p>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R Under development (unstable) (2021-03-18 r80099)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.2 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] ggplot2_3.3.3        magrittr_2.0.1       GenomicRanges_1.43.3</span>
<span class="co">#&gt; [4] GenomeInfoDb_1.27.8  IRanges_2.25.6       S4Vectors_0.29.11   </span>
<span class="co">#&gt; [7] BiocGenerics_0.37.1  memes_0.99.6        </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] bitops_1.0-6                          matrixStats_0.58.0                   </span>
<span class="co">#&gt;  [3] fs_1.5.0                              rprojroot_2.0.2                      </span>
<span class="co">#&gt;  [5] tools_4.1.0                           bslib_0.2.4                          </span>
<span class="co">#&gt;  [7] utf8_1.2.1                            R6_2.5.0                             </span>
<span class="co">#&gt;  [9] colorspace_2.0-0                      withr_2.4.1                          </span>
<span class="co">#&gt; [11] tidyselect_1.1.0                      processx_3.5.0                       </span>
<span class="co">#&gt; [13] compiler_4.1.0                        textshaping_0.3.3                    </span>
<span class="co">#&gt; [15] Biobase_2.51.0                        desc_1.3.0                           </span>
<span class="co">#&gt; [17] DelayedArray_0.17.10                  rtracklayer_1.51.5                   </span>
<span class="co">#&gt; [19] labeling_0.4.2                        sass_0.3.1                           </span>
<span class="co">#&gt; [21] scales_1.1.1                          readr_1.4.0                          </span>
<span class="co">#&gt; [23] pkgdown_1.6.1.9001                    systemfonts_1.0.1                    </span>
<span class="co">#&gt; [25] stringr_1.4.0                         digest_0.6.27                        </span>
<span class="co">#&gt; [27] Rsamtools_2.7.1                       rmarkdown_2.7                        </span>
<span class="co">#&gt; [29] R.utils_2.10.1                        XVector_0.31.1                       </span>
<span class="co">#&gt; [31] pkgconfig_2.0.3                       htmltools_0.5.1.1                    </span>
<span class="co">#&gt; [33] MatrixGenerics_1.3.1                  fastmap_1.1.0                        </span>
<span class="co">#&gt; [35] BSgenome_1.59.2                       highr_0.8                            </span>
<span class="co">#&gt; [37] rlang_0.4.10                          BSgenome.Dmelanogaster.UCSC.dm3_1.4.0</span>
<span class="co">#&gt; [39] jquerylib_0.1.3                       BiocIO_1.1.2                         </span>
<span class="co">#&gt; [41] generics_0.1.0                        farver_2.1.0                         </span>
<span class="co">#&gt; [43] jsonlite_1.7.2                        BiocParallel_1.25.5                  </span>
<span class="co">#&gt; [45] dplyr_1.0.5                           R.oo_1.24.0                          </span>
<span class="co">#&gt; [47] RCurl_1.98-1.3                        GenomeInfoDbData_1.2.4               </span>
<span class="co">#&gt; [49] Matrix_1.3-2                          waldo_0.2.5                          </span>
<span class="co">#&gt; [51] munsell_0.5.0                         fansi_0.4.2                          </span>
<span class="co">#&gt; [53] lifecycle_1.0.0                       R.methodsS3_1.8.1                    </span>
<span class="co">#&gt; [55] stringi_1.5.3                         yaml_2.2.1                           </span>
<span class="co">#&gt; [57] SummarizedExperiment_1.21.1           zlibbioc_1.37.0                      </span>
<span class="co">#&gt; [59] grid_4.1.0                            ggseqlogo_0.1                        </span>
<span class="co">#&gt; [61] crayon_1.4.1                          lattice_0.20-41                      </span>
<span class="co">#&gt; [63] cowplot_1.1.1                         Biostrings_2.59.2                    </span>
<span class="co">#&gt; [65] hms_1.0.0                             knitr_1.31                           </span>
<span class="co">#&gt; [67] ps_1.6.0                              pillar_1.5.1                         </span>
<span class="co">#&gt; [69] rjson_0.2.20                          pkgload_1.2.0                        </span>
<span class="co">#&gt; [71] XML_3.99-0.6                          glue_1.4.2                           </span>
<span class="co">#&gt; [73] evaluate_0.14                         vctrs_0.3.7                          </span>
<span class="co">#&gt; [75] testthat_3.0.2                        gtable_0.3.0                         </span>
<span class="co">#&gt; [77] purrr_0.3.4                           tidyr_1.1.3                          </span>
<span class="co">#&gt; [79] rematch2_2.1.2                        cachem_1.0.4                         </span>
<span class="co">#&gt; [81] xfun_0.22                             restfulr_0.0.13                      </span>
<span class="co">#&gt; [83] ragg_1.1.2                            tibble_3.1.0                         </span>
<span class="co">#&gt; [85] diffobj_0.3.4                         GenomicAlignments_1.27.2             </span>
<span class="co">#&gt; [87] plyranges_1.11.0                      memoise_2.0.0                        </span>
<span class="co">#&gt; [89] cmdfun_1.0.2                          ellipsis_0.3.1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Spencer Nystrom.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
