<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Denovo Motif Discovery Using DREME • memes</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Denovo Motif Discovery Using DREME">
<meta property="og:description" content="memes">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">memes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/install_guide.html">Install MEME</a>
    </li>
    <li>
      <a href="../articles/core_dreme.html">De-Novo Motif Discovery</a>
    </li>
    <li>
      <a href="../articles/core_tomtom.html">Motif Comparison</a>
    </li>
    <li>
      <a href="../articles/core_ame.html">Motif Enrichment</a>
    </li>
    <li>
      <a href="../articles/core_fimo.html">Motif Scanning</a>
    </li>
    <li>
      <a href="../articles/tidy_motifs.html">Motif Manipulation</a>
    </li>
    <li>
      <a href="../articles/integrative_analysis.html">ChIP-Seq Data Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/snystrom/memes/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="core_dreme_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Denovo Motif Discovery Using DREME</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/snystrom/memes/blob/master/vignettes/core_dreme.Rmd" class="external-link"><code>vignettes/core_dreme.Rmd</code></a></small>
      <div class="hidden name"><code>core_dreme.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://snystrom.github.io/memes/" class="external-link">memes</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Verify that memes detects your meme install</span>
<span class="co"># should return all green checks if so.</span>
<span class="fu"><a href="../reference/check_meme_install.html">check_meme_install</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; checking main install</span>
<span class="co">#&gt; ✔ /opt/meme/bin</span>
<span class="co">#&gt; checking util installs</span>
<span class="co">#&gt; ✔ /opt/meme/bin/dreme</span>
<span class="co">#&gt; ✔ /opt/meme/bin/ame</span>
<span class="co">#&gt; ✔ /opt/meme/bin/fimo</span>
<span class="co">#&gt; ✔ /opt/meme/bin/tomtom</span>
<span class="co">#&gt; ✔ /opt/meme/bin/meme</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/fasta_ex/fa1.fa"</span>, package <span class="op">=</span> <span class="st">"memes"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># NOTE: setting e &gt; 1 is usually not recomended. </span>
<span class="co"># the example fasta file only has 1 sequence in it</span>
<span class="co"># to keep the file size low and let the example run quickly.</span>
<span class="co"># I set evalue = 39 because dreme cannot detect high confidence motifs from only 1 sequence.</span>
<span class="va">dreme_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDreme.html">runDreme</a></span><span class="op">(</span><span class="va">fa</span>, <span class="st">"shuffle"</span>, evalue <span class="op">=</span> <span class="fl">39</span>, outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><a href="http://meme-suite.org/doc/dreme.html" class="external-link">DREME Commandline Documentation</a></p>
<div id="aliased-flags" class="section level3">
<h3 class="hasAnchor">
<a href="#aliased-flags" class="anchor"></a>Aliased flags</h3>
<table class="table">
<thead><tr class="header">
<th align="center">memes alias</th>
<th align="center">DREME Flag</th>
<th align="left">description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">nmotifs</td>
<td align="center">m*</td>
<td align="left">max number of motifs to discover</td>
</tr>
<tr class="even">
<td align="center">sec</td>
<td align="center">t</td>
<td align="left">max number of seconds to run</td>
</tr>
<tr class="odd">
<td align="center">evalue</td>
<td align="center">e</td>
<td align="left">max E-value cutoff</td>
</tr>
<tr class="even">
<td align="center">seed</td>
<td align="center">s*</td>
<td align="left">random seed if using “shuffle” as control</td>
</tr>
<tr class="odd">
<td align="center">ngen</td>
<td align="center">g</td>
<td align="left">number of REs to generalize</td>
</tr>
</tbody>
</table>
<p>* flags marked with * must be assigned using their alias</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># equivalent to above</span>
<span class="fu"><a href="../reference/runDreme.html">runDreme</a></span><span class="op">(</span><span class="va">fa</span>, <span class="st">"shuffle"</span>, evalue <span class="op">=</span> <span class="fl">39</span>, outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/runDreme.html">runDreme</a></span><span class="op">(</span><span class="va">fa</span>, <span class="st">"shuffle"</span>, e <span class="op">=</span> <span class="fl">39</span>, outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, nmotifs <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p>dreme results are a <code>data.frame</code>. The <code>motif</code> column contains a <code>universalmotif</code> object with the PCM information for each <em>de-novo</em> discovered motif. This is so that any filtering of the results object also simply filter the available motifs. For more details about each column see the “Value” section of <code><a href="../reference/runDreme.html">?runDreme</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dreme_out</span>
<span class="co">#&gt;                     motif         name altname consensus alphabet strand</span>
<span class="co">#&gt; m01_ATTTATTA &lt;mot:m01_..&gt; m01_ATTTATTA DREME-1  ATTTATTA      DNA     +-</span>
<span class="co">#&gt; m02_AAA      &lt;mot:m02_..&gt;      m02_AAA DREME-2       AAA      DNA     +-</span>
<span class="co">#&gt; m03_AATAATAA &lt;mot:m03_..&gt; m03_AATAATAA DREME-3  AATAATAA      DNA     +-</span>
<span class="co">#&gt; m04_AAT      &lt;mot:m04_..&gt;      m04_AAT DREME-4       AAT      DNA     +-</span>
<span class="co">#&gt; m05_TAA      &lt;mot:m05_..&gt;      m05_TAA DREME-5       TAA      DNA     +-</span>
<span class="co">#&gt;              icscore nsites bkgsites pval eval type                        bkg</span>
<span class="co">#&gt; m01_ATTTATTA      16      4        1  0.5 20.0  PCM 0.667, 0.007, 0.007, 0.319</span>
<span class="co">#&gt; m02_AAA            6      1        1  1.0 24.0  PCM 0.667, 0.007, 0.007, 0.319</span>
<span class="co">#&gt; m03_AATAATAA      16     10        1  0.5  9.0  PCM 0.667, 0.007, 0.007, 0.319</span>
<span class="co">#&gt; m04_AAT            6      2        1  1.0 12.0  PCM 0.667, 0.007, 0.007, 0.319</span>
<span class="co">#&gt; m05_TAA            6      4        1  0.5  0.5  PCM 0.667, 0.007, 0.007, 0.319</span>
<span class="co">#&gt;              rank      seq length positive_hits negative_hits unerased_evalue</span>
<span class="co">#&gt; m01_ATTTATTA    1 ATTTATTA      8             1             0              20</span>
<span class="co">#&gt; m02_AAA         2      AAA      3             1             1              24</span>
<span class="co">#&gt; m03_AATAATAA    3 AATAATAA      8             1             0              18</span>
<span class="co">#&gt; m04_AAT         4      AAT      3             1             1              12</span>
<span class="co">#&gt; m05_TAA         5      TAA      3             1             0               1</span>
<span class="co">#&gt;              positive_total negative_total pos_frac neg_frac</span>
<span class="co">#&gt; m01_ATTTATTA              1              1        1        0</span>
<span class="co">#&gt; m02_AAA                   1              1        1        1</span>
<span class="co">#&gt; m03_AATAATAA              1              1        1        0</span>
<span class="co">#&gt; m04_AAT                   1              1        1        1</span>
<span class="co">#&gt; m05_TAA                   1              1        1        0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [Hidden empty columns: family, organism, qval.]</span></code></pre></div>
<p>The results can be converted back to <code>universalmotif</code> format using <code><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">to_list()</a></code>. The <code><a href="https://rdrr.io/pkg/universalmotif/man/view_motifs.html" class="external-link">view_motifs()</a></code> function accepts a <code>universalmotif</code> list and can be used to visualize the motifs.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/universalmotif/" class="external-link">universalmotif</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>

<span class="va">dreme_out</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">to_list</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/view_motifs.html" class="external-link">view_motifs</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="core_dreme_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>The primary advantage of using the <code>data.frame</code> output allows simple integration with base subsetting, piping, and the <code>tidyverse</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dreme_out</span> <span class="op">%&gt;%</span> 
  <span class="co"># after filtering with dplyr, only motifs with length 3 will be plotted</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">length</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">to_list</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">universalmotif</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/view_motifs.html" class="external-link">view_motifs</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="core_dreme_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p><code>universalmotif</code> manipulations can easily be executed on the motifs as well. For example:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dreme_out</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">to_list</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/merge_motifs.html" class="external-link">merge_motifs</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/view_motifs.html" class="external-link">view_motifs</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="core_dreme_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div id="updating-motif-information" class="section level4">
<h4 class="hasAnchor">
<a href="#updating-motif-information" class="anchor"></a>Updating motif information</h4>
<p>Occasionally, it can be useful to update the metadata associated with a dicovered motif (for example, to assign a new name to a denovo motif). memes provides a few utilities to accomplish this.</p>
<p><code><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">update_motifs()</a></code> will search for specific column names which describe properties of the <code>motif</code> column and update the metadata in the <code>motif</code> column to reflect those values. See <code><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">?update_motifs</a></code> for details.</p>
<p><code>as_universalmotif()</code> will convert one of these special universalmotif data.frames into a universalmotif list after updating the metadata to reflect values as in <code><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">update_motifs()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># update_motifs will update the values in the motif column</span>
<span class="co"># to values in the data.frame</span>
<span class="va">dreme_edit</span> <span class="op">&lt;-</span> <span class="va">dreme_out</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"one"</span>, <span class="st">"two"</span>, <span class="st">"three"</span>, <span class="st">"four"</span>, <span class="st">"five"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">update_motifs</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># to_list() will first update motif information</span>
<span class="co"># before returning only the motif column</span>
<span class="va">edit_motifs</span> <span class="op">&lt;-</span> <span class="va">dreme_out</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"one"</span>, <span class="st">"two"</span>, <span class="st">"three"</span>, <span class="st">"four"</span>, <span class="st">"five"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">to_list</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># The following outputs are identical</span>
<span class="co"># where edit_motifs is a list of motifs</span>
<span class="co"># and dreme_edit is a data.frame with a motif list column</span>
<span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">edit_motifs</span><span class="op">$</span><span class="va">one</span>, <span class="va">dreme_edit</span><span class="op">$</span><span class="va">motif</span><span class="op">$</span><span class="va">one</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
</div>
<div id="notes-about-shuffled-control-sequences" class="section level3">
<h3 class="hasAnchor">
<a href="#notes-about-shuffled-control-sequences" class="anchor"></a>Notes about shuffled control sequences</h3>
<p>Setting <code>control = "shuffle"</code> will use dreme’s random number generator to shuffle the input sequences. By default, dreme will use <code>1</code> as the random seed, so repeat runs of the same shuffle command will produce the same output. To change the random seed, pass <code>seed = [your random seed]</code> to <code><a href="../reference/runDreme.html">runDreme()</a></code>. <strong>NOTE:</strong> beware system-specific differences. As of MEME v5, dreme will compile using the default python installation on a system (either python2.7 or python3). The random number generator changed between python2.7 and python3, so results will not be reproducible between systems using python2.7 vs 3 <strong>even if setting the same random seed</strong>.</p>
<p>One way to overcome this is to manually shuffle the sequences within R. This can be done easily using <code><a href="https://rdrr.io/pkg/universalmotif/man/shuffle_sequences.html" class="external-link">universalmotif::shuffle_sequences()</a></code>. Set <code>k = 2</code> to preserve dinucleotide frequency (similar to dreme’s built-in shuffle), and set <code>rng.seed</code> to any number to create a reproducible shuffle. The output of this function can be used directly as the control sequences.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create random sequences to use for this example</span>
<span class="va">seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/create_sequences.html" class="external-link">create_sequences</a></span><span class="op">(</span>rng.seed <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="co"># Shuffle sequences preserving dinucleotide frequency</span>
<span class="va">shuffle</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/shuffle_sequences.html" class="external-link">shuffle_sequences</a></span><span class="op">(</span><span class="va">seq</span>, k <span class="op">=</span> <span class="fl">2</span>, rng.seed <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></code></pre></div>
</div>
<div id="analysis-on-multiple-groups-and-differential-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#analysis-on-multiple-groups-and-differential-analysis" class="anchor"></a>Analysis on Multiple Groups and Differential Analysis</h3>
<p>Often, users want to perform motif analysis on many groups of sequences. For example, here we have ChIP-seq peaks for a transcription factor, E93. Analysis of chromatin accessibility in E93 peaks revealed sites that Increase accessibility, Decrease accessibility, or remain Static following E93 binding.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_chip_summits"</span>, package <span class="op">=</span> <span class="st">"memes"</span><span class="op">)</span>
<span class="va">peaks</span> <span class="op">&lt;-</span> <span class="va">example_chip_summits</span></code></pre></div>
<p>To examine whether there are differences in motif content between increasing, decreasing, and static sites, we split the peaks into a list by their response to E93.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">by_behavior</span> <span class="op">&lt;-</span> <span class="va">peaks</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-anchor.html" class="external-link">anchor_center</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">mutate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">$</span><span class="va">e93_sensitive_behavior</span><span class="op">)</span></code></pre></div>
<p>Next, this list can be used directly in <code>get_sequences()</code> to generate a list of sequences for each set of peaks.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dm.genome</span> <span class="op">&lt;-</span> <span class="fu">BSgenome.Dmelanogaster.UCSC.dm3</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/BSgenome.Dmelanogaster.UCSC.dm3/man/package.html" class="external-link">BSgenome.Dmelanogaster.UCSC.dm3</a></span>

<span class="va">seq_by_behavior</span> <span class="op">&lt;-</span> <span class="va">by_behavior</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/get_sequence.html">get_sequence</a></span><span class="op">(</span><span class="va">dm.genome</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">seq_by_behavior</span><span class="op">)</span>
<span class="co">#&gt; [1] "Decreasing" "Increasing" "Static"</span></code></pre></div>
<p>To run DREME on each set using shuffled input sequence as background, run:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/runDreme.html">runDreme</a></span><span class="op">(</span><span class="va">seq_by_behavior</span>, control <span class="op">=</span> <span class="st">"shuffle"</span><span class="op">)</span></code></pre></div>
<div id="discriminative-analysis-using-list-input" class="section level4">
<h4 class="hasAnchor">
<a href="#discriminative-analysis-using-list-input" class="anchor"></a>Discriminative analysis using list input</h4>
<p>For this analysis, however, we are most interested in identifying motifs associated with increasing and decreasing that do not involve E93 binding. Therefore, a more appropriate control is to use the Static sites as background.</p>
<p>As always, an <code>XStringSet</code> object can be used as the control regions. However, running dreme in this way will run 3 jobs:</p>
<ol style="list-style-type: decimal">
<li>Increasing vs Static</li>
<li>Decreasing vs Static</li>
<li>Static vs Static</li>
</ol>
<p>This will waste time, as job #3 will detect no motifs (since input &amp; control are identical), but will still take a long time to run. <code><a href="../reference/runDreme.html">runDreme()</a></code> has additional functionality to help avoid these issues, and to facilitate more complicated analysis designs.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/runDreme.html">runDreme</a></span><span class="op">(</span><span class="va">seq_by_behavior</span>, control <span class="op">=</span> <span class="va">seq_by_behavior</span><span class="op">$</span><span class="va">Static</span><span class="op">)</span></code></pre></div>
<p>If the input to <code>runDreme</code> is a named list of <code>XStringSet</code> objects, <code>control</code> can be set to one or more values from <code><a href="https://rdrr.io/r/base/names.html" class="external-link">names(input)</a></code> to use those regions as background. It will skip running those regions as the input. The following code will result in these comparisons:</p>
<ol style="list-style-type: decimal">
<li>Increasing vs Static</li>
<li>Decreasing vs Static</li>
</ol>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/runDreme.html">runDreme</a></span><span class="op">(</span><span class="va">seq_by_behavior</span>, control <span class="op">=</span> <span class="st">"Static"</span><span class="op">)</span></code></pre></div>
<p>If multiple names are used in the <code>control</code> section, they will be combined together to make a single control set which will be used for all comparisons. Here, we use “Static” and “Decreasing” sites as the control, which will result in only running 1 comparison: Increasing vs Static+Decreasing.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/runDreme.html">runDreme</a></span><span class="op">(</span><span class="va">seq_by_behavior</span>, control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Static"</span>, <span class="st">"Decreasing"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="importing-previous-data" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-previous-data" class="anchor"></a>Importing previous data</h2>
<p><code><a href="../reference/importDremeXML.html">importDremeXML()</a></code> can be used to import a <code>dreme.xml</code> file from a previous run on the MEME server or on the commandline. Details for how to save data from the DREME webserver are below.</p>
<div id="saving-data-from-dreme-web-server" class="section level3">
<h3 class="hasAnchor">
<a href="#saving-data-from-dreme-web-server" class="anchor"></a>Saving data from DREME Web Server</h3>
<p>To download XML data from the MEME Server, right-click the DREME XML output link and “Save Target As” or “Save Link As” (see example image below), and save as <code>&lt;filename&gt;.xml</code>. This file can be read using <code><a href="../reference/importDremeXML.html">importDremeXML()</a></code></p>
<p><img src="save_dreme.png"></p>
</div>
</div>
<div id="citation" class="section level1">
<h1 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h1>
<p>memes is a wrapper for a select few tools from the MEME Suite, which were developed by another group. In addition to citing memes, please cite the MEME Suite tools corresponding to the tools you use.</p>
<p>If you use <code><a href="../reference/runDreme.html">runDreme()</a></code> in your analysis, please cite:</p>
<p>Timothy L. Bailey, “DREME: Motif discovery in transcription factor ChIP-seq data”, Bioinformatics, 27(12):1653-1659, 2011. <a href="https://academic.oup.com/bioinformatics/article/27/12/1653/257754" class="external-link">full text</a></p>
<div id="licensing-restrictions" class="section level2">
<h2 class="hasAnchor">
<a href="#licensing-restrictions" class="anchor"></a>Licensing Restrictions</h2>
<p>The MEME Suite is free for non-profit use, but for-profit users should purchase a license. See the <a href="http://meme-suite.org/doc/copyright.html" class="external-link">MEME Suite Copyright Page</a> for details.</p>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R Under development (unstable) (2021-04-05 r80145)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.2 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] plyranges_1.11.0      GenomicRanges_1.43.4  GenomeInfoDb_1.27.10 </span>
<span class="co">#&gt; [4] IRanges_2.25.6        S4Vectors_0.29.13     BiocGenerics_0.37.1  </span>
<span class="co">#&gt; [7] dplyr_1.0.5           universalmotif_1.9.46 memes_0.99.9         </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] bitops_1.0-6                          matrixStats_0.58.0                   </span>
<span class="co">#&gt;  [3] fs_1.5.0                              usethis_2.0.1                        </span>
<span class="co">#&gt;  [5] rprojroot_2.0.2                       tools_4.1.0                          </span>
<span class="co">#&gt;  [7] bslib_0.2.4                           utf8_1.2.1                           </span>
<span class="co">#&gt;  [9] R6_2.5.0                              colorspace_2.0-0                     </span>
<span class="co">#&gt; [11] withr_2.4.1                           tidyselect_1.1.0                     </span>
<span class="co">#&gt; [13] processx_3.5.1                        compiler_4.1.0                       </span>
<span class="co">#&gt; [15] textshaping_0.3.3                     cli_2.4.0                            </span>
<span class="co">#&gt; [17] Biobase_2.51.0                        xml2_1.3.2                           </span>
<span class="co">#&gt; [19] desc_1.3.0                            DelayedArray_0.17.10                 </span>
<span class="co">#&gt; [21] rtracklayer_1.51.5                    sass_0.3.1                           </span>
<span class="co">#&gt; [23] scales_1.1.1                          readr_1.4.0                          </span>
<span class="co">#&gt; [25] pkgdown_1.6.1.9001                    Rsamtools_2.7.1                      </span>
<span class="co">#&gt; [27] systemfonts_1.0.1                     stringr_1.4.0                        </span>
<span class="co">#&gt; [29] digest_0.6.27                         rmarkdown_2.7                        </span>
<span class="co">#&gt; [31] R.utils_2.10.1                        XVector_0.31.1                       </span>
<span class="co">#&gt; [33] pkgconfig_2.0.3                       htmltools_0.5.1.1                    </span>
<span class="co">#&gt; [35] MatrixGenerics_1.3.1                  BSgenome_1.59.2                      </span>
<span class="co">#&gt; [37] fastmap_1.1.0                         highr_0.8                            </span>
<span class="co">#&gt; [39] rlang_0.4.10                          rstudioapi_0.13                      </span>
<span class="co">#&gt; [41] BSgenome.Dmelanogaster.UCSC.dm3_1.4.0 BiocIO_1.1.2                         </span>
<span class="co">#&gt; [43] jquerylib_0.1.3                       generics_0.1.0                       </span>
<span class="co">#&gt; [45] farver_2.1.0                          jsonlite_1.7.2                       </span>
<span class="co">#&gt; [47] BiocParallel_1.25.5                   R.oo_1.24.0                          </span>
<span class="co">#&gt; [49] RCurl_1.98-1.3                        magrittr_2.0.1                       </span>
<span class="co">#&gt; [51] GenomeInfoDbData_1.2.4                Matrix_1.3-2                         </span>
<span class="co">#&gt; [53] Rcpp_1.0.6                            waldo_0.2.5                          </span>
<span class="co">#&gt; [55] munsell_0.5.0                         fansi_0.4.2                          </span>
<span class="co">#&gt; [57] lifecycle_1.0.0                       R.methodsS3_1.8.1                    </span>
<span class="co">#&gt; [59] stringi_1.5.3                         yaml_2.2.1                           </span>
<span class="co">#&gt; [61] MASS_7.3-53.1                         SummarizedExperiment_1.21.2          </span>
<span class="co">#&gt; [63] zlibbioc_1.37.0                       grid_4.1.0                           </span>
<span class="co">#&gt; [65] ggseqlogo_0.1                         crayon_1.4.1                         </span>
<span class="co">#&gt; [67] lattice_0.20-41                       Biostrings_2.59.2                    </span>
<span class="co">#&gt; [69] hms_1.0.0                             knitr_1.31                           </span>
<span class="co">#&gt; [71] ps_1.6.0                              pillar_1.5.1                         </span>
<span class="co">#&gt; [73] rjson_0.2.20                          pkgload_1.2.1                        </span>
<span class="co">#&gt; [75] XML_3.99-0.6                          glue_1.4.2                           </span>
<span class="co">#&gt; [77] evaluate_0.14                         vctrs_0.3.7                          </span>
<span class="co">#&gt; [79] testthat_3.0.2                        gtable_0.3.0                         </span>
<span class="co">#&gt; [81] purrr_0.3.4                           tidyr_1.1.3                          </span>
<span class="co">#&gt; [83] rematch2_2.1.2                        cachem_1.0.4                         </span>
<span class="co">#&gt; [85] ggplot2_3.3.3                         xfun_0.22                            </span>
<span class="co">#&gt; [87] restfulr_0.0.13                       ragg_1.1.2                           </span>
<span class="co">#&gt; [89] tibble_3.1.0                          diffobj_0.3.4                        </span>
<span class="co">#&gt; [91] GenomicAlignments_1.27.2              memoise_2.0.0                        </span>
<span class="co">#&gt; [93] cmdfun_1.0.2                          ellipsis_0.3.1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Spencer Nystrom.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
