<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tidying Motif Metadata • memes</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tidying Motif Metadata">
<meta property="og:description" content="memes">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">memes</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/install_guide.html">Install MEME</a>
    </li>
    <li>
      <a href="../articles/core_dreme.html">De-Novo Motif Discovery</a>
    </li>
    <li>
      <a href="../articles/core_tomtom.html">Motif Comparison</a>
    </li>
    <li>
      <a href="../articles/core_ame.html">Motif Enrichment</a>
    </li>
    <li>
      <a href="../articles/core_fimo.html">Motif Scanning</a>
    </li>
    <li>
      <a href="../articles/tidy_motifs.html">Motif Manipulation</a>
    </li>
    <li>
      <a href="../articles/integrative_analysis.html">ChIP-Seq Data Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/snystrom/memes/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tidying Motif Metadata</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/snystrom/memes/blob/HEAD/vignettes/tidy_motifs.Rmd" class="external-link"><code>vignettes/tidy_motifs.Rmd</code></a></small>
      <div class="hidden name"><code>tidy_motifs.Rmd</code></div>

    </div>

    
    
<p>memes uses the <code>universalmotif</code> package to simplify working with motif metadata. <code>universalmotif</code> objects can be represented in an alternative form, the <code>unviversalmotif_df</code> which allows users to manipulate motif metadata just as they would a normal R data.frame (in fact, they are just R data.frames).</p>
<p>These objects are useful for tidying motif metadata to prepare a motif database for use with memes, or performing any other data-driven tasks involving motifs. Here I describe one way these data structures can be used to construct a motif database for use with memes.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://snystrom.github.io/memes/" class="external-link">memes</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/universalmotif/" class="external-link">universalmotif</a></span><span class="op">)</span></code></pre></div>
<p>The <code>MotifDb</code> package makes it easy to query thousands of motifs from public databases. Here, I will describe how to use one of these queries as input to memes functions, and how to manipulate the resulting motifs to prepare them for MEME suite tools.</p>
<p>I will use the motifs from the <a href="https://mccb.umassmed.edu/ffs/" class="external-link">FlyFactorSurvey</a> as an example. They can be accessed from <code>MotifDb</code> using the following query.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactorDb</span> <span class="op">&lt;-</span> <span class="fu">MotifDb</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/MotifDb/man/MotifDb.html" class="external-link">MotifDb</a></span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">MotifDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MotifDb/man/query.html" class="external-link">query</a></span><span class="op">(</span><span class="st">"FlyFactorSurvey"</span><span class="op">)</span>
<span class="co">#&gt; See system.file("LICENSE", package="MotifDb") for use restrictions.</span></code></pre></div>
<p>Use <code><a href="https://rdrr.io/pkg/universalmotif/man/convert_motifs.html" class="external-link">universalmotif::convert_motifs()</a></code> to convert a <code>MotifDb</code> query into motif objects. In many cases, the resulting list can be used directly as input to memes functions, like <code>runTomTom</code>, or <code>runAme</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactorMotifs</span> <span class="op">&lt;-</span> <span class="va">flyFactorDb</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/convert_motifs.html" class="external-link">convert_motifs</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>But there are some issues with this database. For example, the following motif name is the FlyBase gene number, and the alternate name is the actual informative name of the PWM. The MEME Suite relies more heavily on the primary name, so it would be nice if the database used interpretable names.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactorMotifs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;        Motif name:   FBgn0259750</span>
<span class="co">#&gt;    Alternate name:   ab_SANGER_10_FBgn0259750</span>
<span class="co">#&gt;          Organism:   Dmelanogaster</span>
<span class="co">#&gt;          Alphabet:   DNA</span>
<span class="co">#&gt;              Type:   PPM</span>
<span class="co">#&gt;           Strands:   +-</span>
<span class="co">#&gt;          Total IC:   15.01</span>
<span class="co">#&gt;       Pseudocount:   1</span>
<span class="co">#&gt;         Consensus:   BWNRCCAGGWMCNNTNNNGNN</span>
<span class="co">#&gt;      Target sites:   20</span>
<span class="co">#&gt;        Extra info:   [dataSource] FlyFactorSurvey </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     B    W    N    R C C A G G    W    M    C    N    N    T    N    N    N</span>
<span class="co">#&gt; A 0.0 0.50 0.20 0.35 0 0 1 0 0 0.55 0.35 0.05 0.20 0.45 0.20 0.10 0.40 0.40</span>
<span class="co">#&gt; C 0.3 0.15 0.25 0.00 1 1 0 0 0 0.10 0.65 0.70 0.45 0.25 0.10 0.25 0.25 0.10</span>
<span class="co">#&gt; G 0.4 0.05 0.50 0.65 0 0 0 1 1 0.00 0.00 0.05 0.05 0.15 0.05 0.20 0.05 0.15</span>
<span class="co">#&gt; T 0.3 0.30 0.05 0.00 0 0 0 0 0 0.35 0.00 0.20 0.30 0.15 0.65 0.45 0.30 0.35</span>
<span class="co">#&gt;      G    N    N</span>
<span class="co">#&gt; A 0.25 0.50 0.30</span>
<span class="co">#&gt; C 0.10 0.25 0.25</span>
<span class="co">#&gt; G 0.55 0.15 0.45</span>
<span class="co">#&gt; T 0.10 0.10 0.00</span></code></pre></div>
<p>The universalmotif function <code><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">to_df()</a></code> converts universalmotif lists into <code>universalmotif_df</code> format which can be used to update motif entries. This is particularly useful when dealing with several motifs at once.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op">&lt;-</span> <span class="va">flyFactorMotifs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">to_df</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>The columns of the <code>universalmotif_df</code> can be freely changed to edit the properties of the motifs stored in the <code>motif</code> column. Just like standard data.frames, additional columns can be added to store additional metadata. For more details on these objects see the help page: <code><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">?universalmotif::to_df</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># The following columns can be changed to update motif metadata</span>
<span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="va">names</span>
<span class="co">#&gt;  [1] "motif"       "name"        "altname"     "family"      "organism"   </span>
<span class="co">#&gt;  [6] "consensus"   "alphabet"    "strand"      "icscore"     "nsites"     </span>
<span class="co">#&gt; [11] "bkgsites"    "pval"        "qval"        "eval"        "type"       </span>
<span class="co">#&gt; [16] "pseudocount" "bkg"         "dataSource"</span></code></pre></div>
<p>using the <code>universalmotif_df</code>, we can quickly see that the issue with FBgn numbers only applies to certain entries. And TFs which are represented by multiple motifs in the database are assigned the same name. The MEME Suite tools which use a motif database (like TomTom and AME) require that the entries have unique primary identifiers, therefore the default names will not be appropriate.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span>
<span class="co">#&gt;          motif        name                  altname      organism</span>
<span class="co">#&gt; 1 &lt;mot:FBgn..&gt; FBgn0259750 ab_SANGER_10_FBgn0259750 Dmelanogaster</span>
<span class="co">#&gt; 2 &lt;mot:FBgn..&gt; FBgn0259750  ab_SOLEXA_5_FBgn0259750 Dmelanogaster</span>
<span class="co">#&gt; 3  &lt;mot:abd-A&gt;       abd-A Abd-A_FlyReg_FBgn0000014 Dmelanogaster</span>
<span class="co">#&gt; 4  &lt;mot:Abd-B&gt;       Abd-B Abd-B_FlyReg_FBgn0000015 Dmelanogaster</span>
<span class="co">#&gt; 5  &lt;mot:abd-A&gt;       abd-A    AbdA_Cell_FBgn0000014 Dmelanogaster</span>
<span class="co">#&gt;                    consensus alphabet strand   icscore nsites type pseudocount</span>
<span class="co">#&gt; 1      BWNRCCAGGWMCNNTNNNGNN      DNA     +- 15.013759     20  PPM           1</span>
<span class="co">#&gt; 2 NNNNHNRCCAGGWMCYNNNNNNNNNN      DNA     +- 14.570384    446  PPM           1</span>
<span class="co">#&gt; 3                   KNMATWAW      DNA     +-  7.371641     37  PPM           1</span>
<span class="co">#&gt; 4                   YCATAAAA      DNA     +-  7.831193      7  PPM           1</span>
<span class="co">#&gt; 5                    TTAATKA      DNA     +-  8.976569     18  PPM           1</span>
<span class="co">#&gt;                      bkg      dataSource</span>
<span class="co">#&gt; 1 0.25, 0.25, 0.25, 0.25 FlyFactorSurvey</span>
<span class="co">#&gt; 2 0.25, 0.25, 0.25, 0.25 FlyFactorSurvey</span>
<span class="co">#&gt; 3 0.25, 0.25, 0.25, 0.25 FlyFactorSurvey</span>
<span class="co">#&gt; 4 0.25, 0.25, 0.25, 0.25 FlyFactorSurvey</span>
<span class="co">#&gt; 5 0.25, 0.25, 0.25, 0.25 FlyFactorSurvey</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [Hidden empty columns: family, bkgsites, pval, qval, eval.]</span></code></pre></div>
<p>However, the <code>altname</code> slots from the <code>motifDb</code> query are already unique, so we can make them the primary name.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">flyFactor_data</span><span class="op">$</span><span class="va">altname</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">flyFactor_data</span><span class="op">$</span><span class="va">altname</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
<p>An easy way is to use <code><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">dplyr::rename</a></code> to swap the columns.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span><span class="st">"altname"</span> <span class="op">=</span> <span class="st">"name"</span>, 
                <span class="st">"name"</span> <span class="op">=</span> <span class="st">"altname"</span><span class="op">)</span></code></pre></div>
<p>The <code>name</code> column now contains the full motif name.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt;            motif     altname                     name      organism</span>
<span class="co">#&gt; 1 * &lt;mot:FBgn..&gt; FBgn0259750 ab_SANGER_10_FBgn0259750 Dmelanogaster</span>
<span class="co">#&gt; 2 * &lt;mot:FBgn..&gt; FBgn0259750  ab_SOLEXA_5_FBgn0259750 Dmelanogaster</span>
<span class="co">#&gt; 3 *  &lt;mot:abd-A&gt;       abd-A Abd-A_FlyReg_FBgn0000014 Dmelanogaster</span>
<span class="co">#&gt;                    consensus alphabet strand   icscore nsites type pseudocount</span>
<span class="co">#&gt; 1      BWNRCCAGGWMCNNTNNNGNN      DNA     +- 15.013759     20  PPM           1</span>
<span class="co">#&gt; 2 NNNNHNRCCAGGWMCYNNNNNNNNNN      DNA     +- 14.570384    446  PPM           1</span>
<span class="co">#&gt; 3                   KNMATWAW      DNA     +-  7.371641     37  PPM           1</span>
<span class="co">#&gt;                      bkg      dataSource</span>
<span class="co">#&gt; 1 0.25, 0.25, 0.25, 0.25 FlyFactorSurvey</span>
<span class="co">#&gt; 2 0.25, 0.25, 0.25, 0.25 FlyFactorSurvey</span>
<span class="co">#&gt; 3 0.25, 0.25, 0.25, 0.25 FlyFactorSurvey</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [Hidden empty columns: family, bkgsites, pval, qval, eval.]</span>
<span class="co">#&gt; [Rows marked with * are changed. Run update_motifs() or to_list() to</span>
<span class="co">#&gt;   apply changes.]</span></code></pre></div>
<p>Next to solve the issue with the FBgn’s. FBgn numbers are unique identifiers for a gene within a given FlyBase reference assembly. However, FBgn numbers are not stable over time (i.e. the same gene may have a different FBgn number between reference assemblies), therefore they are unreliable values to determine the correct gene symbol. <a href="https://flybase.org" class="external-link">FlyBase.org</a> has a nice <a href="https://flybase.org/convert/id" class="external-link">conversion tool</a> which can be used to update FBgn numbers.</p>
<p>As of this writing in March 2021, the FBgn entries provided by the Fly Factor Survey database are out of date. In order to demonstrate an example of methods for tidying motif metadata, I won’t use the FlyBase conversion tool, but will instead highlight some approaches which may be more generally useful when working with motif databases from disparate sources.</p>
<p>For this example, we will try to grab the correct gene name from the motif name, which is stored in the first field of the name, formatted as follows: “<gene><em><sequencing-platform></sequencing-platform></em><fbgn>”.</fbgn></gene></p>
<p>We use <code><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">tidyr::separate</a></code> to split out the first entry to the <code>tifd</code> column, then only use this value if the altname contains an FBgn.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> 
  <span class="co"># Critical to set remove = FALSE to keep the `name` column</span>
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">name</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tfid"</span><span class="op">)</span>, remove <span class="op">=</span> <span class="cn">FALSE</span>, extra <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="co"># Only use the tfid if the altname contains an FBgn</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>altname <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^FBgn"</span>, <span class="va">altname</span><span class="op">)</span>, <span class="va">tfid</span>, <span class="va">altname</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Now, the first two entries are listed as “ab” instead of “FBgn0259750”.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt;            motif altname                     name tfid      organism</span>
<span class="co">#&gt; 1 * &lt;mot:FBgn..&gt;      ab ab_SANGER_10_FBgn0259750   ab Dmelanogaster</span>
<span class="co">#&gt; 2 * &lt;mot:FBgn..&gt;      ab  ab_SOLEXA_5_FBgn0259750   ab Dmelanogaster</span>
<span class="co">#&gt; 3 *  &lt;mot:abd-A&gt;   abd-A Abd-A_FlyReg_FBgn0000014  Abd Dmelanogaster</span>
<span class="co">#&gt;                    consensus alphabet strand   icscore nsites type pseudocount</span>
<span class="co">#&gt; 1      BWNRCCAGGWMCNNTNNNGNN      DNA     +- 15.013759     20  PPM           1</span>
<span class="co">#&gt; 2 NNNNHNRCCAGGWMCYNNNNNNNNNN      DNA     +- 14.570384    446  PPM           1</span>
<span class="co">#&gt; 3                   KNMATWAW      DNA     +-  7.371641     37  PPM           1</span>
<span class="co">#&gt;                      bkg      dataSource</span>
<span class="co">#&gt; 1 0.25, 0.25, 0.25, 0.25 FlyFactorSurvey</span>
<span class="co">#&gt; 2 0.25, 0.25, 0.25, 0.25 FlyFactorSurvey</span>
<span class="co">#&gt; 3 0.25, 0.25, 0.25, 0.25 FlyFactorSurvey</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [Hidden empty columns: family, bkgsites, pval, qval, eval.]</span>
<span class="co">#&gt; [Rows marked with * are changed. Run update_motifs() or to_list() to</span>
<span class="co">#&gt;   apply changes.]</span></code></pre></div>
<p>Next, because the FBgn’s are out of date, we will remove them from the “names” to shorten up the motif names. This also makes the motif name more comparable to the original motif names from the <a href="https://mccb.umassmed.edu/ffs/" class="external-link">FlyFactor Survey</a>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_FBgn\\d+"</span>, <span class="st">""</span>, <span class="va">name</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="a-few-reality-checks">A few reality checks<a class="anchor" aria-label="anchor" href="#a-few-reality-checks"></a>
</h2>
<p>It’s worth taking a look at the instances where the <code>altname</code> and our parsed <code>tfid</code> do not match. This is a good way to ensure we haven’t missed any important edge cases in the data. As new edge cases are encountered, we can develop new rules for tidying the data to ensure a high quality set of motifs.</p>
<p>Start by simply filtering for all instance where there is a mismatch between <code>altname</code> and <code>tfid</code>.</p>
<p>Carefully compare the <code>altname</code>, <code>name</code>, and <code>tfid</code> columns. Why might the values differ? Are there instances that make you question the data?</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">altname</span> <span class="op">!=</span> <span class="va">tfid</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="co"># I'm only showing the first 5 rows for brevity, but take a look at the full</span>
  <span class="co"># data and see what patterns you notice</span>
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span>
<span class="co">#&gt;           motif altname         name tfid      organism consensus alphabet</span>
<span class="co">#&gt; 1 * &lt;mot:abd-A&gt;   abd-A Abd-A_FlyReg  Abd Dmelanogaster  KNMATWAW      DNA</span>
<span class="co">#&gt; 2 * &lt;mot:Abd-B&gt;   Abd-B Abd-B_FlyReg  Abd Dmelanogaster  YCATAAAA      DNA</span>
<span class="co">#&gt; 3 * &lt;mot:abd-A&gt;   abd-A    AbdA_Cell AbdA Dmelanogaster   TTAATKA      DNA</span>
<span class="co">#&gt; 4 * &lt;mot:abd-A&gt;   abd-A  AbdA_SOLEXA AbdA Dmelanogaster  NTTAATKR      DNA</span>
<span class="co">#&gt; 5 * &lt;mot:Abd-B&gt;   Abd-B    AbdB_Cell AbdB Dmelanogaster GWTTTATKA      DNA</span>
<span class="co">#&gt;   strand  icscore nsites type pseudocount                    bkg</span>
<span class="co">#&gt; 1     +- 7.371641     37  PPM           1 0.25, 0.25, 0.25, 0.25</span>
<span class="co">#&gt; 2     +- 7.831193      7  PPM           1 0.25, 0.25, 0.25, 0.25</span>
<span class="co">#&gt; 3     +- 8.976569     18  PPM           1 0.25, 0.25, 0.25, 0.25</span>
<span class="co">#&gt; 4     +- 8.330000    662  PPM           1 0.25, 0.25, 0.25, 0.25</span>
<span class="co">#&gt; 5     +- 9.147674     21  PPM           1 0.25, 0.25, 0.25, 0.25</span>
<span class="co">#&gt;        dataSource</span>
<span class="co">#&gt; 1 FlyFactorSurvey</span>
<span class="co">#&gt; 2 FlyFactorSurvey</span>
<span class="co">#&gt; 3 FlyFactorSurvey</span>
<span class="co">#&gt; 4 FlyFactorSurvey</span>
<span class="co">#&gt; 5 FlyFactorSurvey</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [Hidden empty columns: family, bkgsites, pval, qval, eval.]</span>
<span class="co">#&gt; [Rows marked with * are changed. Run update_motifs() or to_list() to</span>
<span class="co">#&gt;   apply changes.]</span></code></pre></div>
<p>One thing that becomes obvious is that many motifs have mismatched <code>altname</code>/<code>tfid</code> values because of capitalization or hyphenation differences. You can use domain-specific knowledge to assess which one is correct. For <em>Drosophila</em>, “abd-A” is correct over “AbdA”, for example.</p>
<p>After manually inspecting these rows, I determined that instances of different capitalization, hyphenation, or names that contain “.” or “()” can be ignored. To further investigate the data, I will ignore capitalization and special character differences as follows:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="co"># calling tolower() on both columns removes capitalization as a difference</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">altname</span><span class="op">)</span> <span class="op">!=</span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">tfid</span><span class="op">)</span>,
                <span class="co"># Select all altnames that do not contain "-", "." or "("</span>
                <span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"-|\\.|\\("</span>, <span class="va">altname</span><span class="op">)</span>,
                <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="co"># I'll visalize only these columns for brevity</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">altname</span>, <span class="va">tfid</span>, <span class="va">name</span>, <span class="va">consensus</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>
<span class="co">#&gt;    altname    tfid              name            consensus</span>
<span class="co">#&gt; 1       da      ac    ac_da_SANGER_5             RCACCTGC</span>
<span class="co">#&gt; 2       da    amos amos_da_SANGER_10          RMCAYMTGBCV</span>
<span class="co">#&gt; 3       da     ase  ase_da_SANGER_10              CACCTGY</span>
<span class="co">#&gt; 4       da     ato  ato_da_SANGER_10          MCAYMTGNCRC</span>
<span class="co">#&gt; 5       da     ato ato_da_SANGER_5_2            GNCAKRTGN</span>
<span class="co">#&gt; 6       da     ato ato_da_SANGER_5_3            MCAYMTGNC</span>
<span class="co">#&gt; 7       da    cato cato_da_SANGER_10           MRCANMTGWC</span>
<span class="co">#&gt; 8      Zif CG10267  CG10267_SANGER_5           WNYAACACTR</span>
<span class="co">#&gt; 9      Zif CG10267  CG10267_SOLEXA_5 CNNNNWAYAACACTASNMNN</span>
<span class="co">#&gt; 10    mamo CG11071  CG11071_SANGER_5         YMMGCCTAYNNM</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [Note: incomplete universalmotif_df object.]</span></code></pre></div>
<p>Next, what is obvious is that several <code>altnames</code> set to “da” have a high number of mismatched <code>tfid</code>s. For instance, <code>amos_da_SANGER_10</code>. When checking the <a href="https://mccb.umassmed.edu/ffs/TFdetails.php?FlybaseID=FBgn0000413" class="external-link">FlyFactorSurvey page for da</a>, it reveals only 1 motif corresponds to this factor. Checking the <a href="https://mccb.umassmed.edu/ffs/TFdetails.php?FlybaseID=FBgn0003270" class="external-link">page for amos</a> shows a match to <code>amos_da_SANGER_10</code>. Therefore, we can conclude that factors assigned the name of <code>da</code> are incorrectly assigned, and we should prefer our parsed <code>tfid</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> 
  <span class="co"># rename all "da" instances using their tfid value instead</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>altname <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">altname</span> <span class="op">==</span> <span class="st">"da"</span>, <span class="va">tfid</span>, <span class="va">altname</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Now we’ve handled the “da” mismatches, we filter them out to identify new special cases.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">altname</span><span class="op">)</span> <span class="op">!=</span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">tfid</span><span class="op">)</span>,
                <span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"-|\\.|\\("</span>, <span class="va">altname</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">altname</span>, <span class="va">tfid</span>, <span class="va">name</span>, <span class="va">consensus</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>
<span class="co">#&gt;    altname    tfid             name            consensus</span>
<span class="co">#&gt; 1      Zif CG10267 CG10267_SANGER_5           WNYAACACTR</span>
<span class="co">#&gt; 2      Zif CG10267 CG10267_SOLEXA_5 CNNNNWAYAACACTASNMNN</span>
<span class="co">#&gt; 3     mamo CG11071 CG11071_SANGER_5         YMMGCCTAYNNM</span>
<span class="co">#&gt; 4      lms CG13424     CG13424_Cell             NYTAATTR</span>
<span class="co">#&gt; 5      lms CG13424 CG13424_SOLEXA_2           NNYTAATTRN</span>
<span class="co">#&gt; 6      lms CG13424   CG13424_SOLEXA              YTAATTR</span>
<span class="co">#&gt; 7      erm CG31670 CG31670_SANGER_5          AAAWGMGCAWC</span>
<span class="co">#&gt; 8      erm CG31670 CG31670_SOLEXA_5      NNAAAWGAGCAAYNV</span>
<span class="co">#&gt; 9     Spps  CG5669 CG5669_SANGER_10      DRKGGGCGKGGCCAM</span>
<span class="co">#&gt; 10    Spps  CG5669  CG5669_SOLEXA_5      NGKGGGCGKGGCNWN</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [Note: incomplete universalmotif_df object.]</span></code></pre></div>
<p>The next thing to notice about these data is that entries with “CG” prefixed tfids are often mismatched. This is because when the FlyFactor survey was conducted, many genes were unnamed, and thus assigned a CG from FlyBase. As time has gone on, some CG’s have been named. Checking the <a href="http://flybase.org/reports/FBgn0037446" class="external-link">FlyBase page for CG10267</a> reveals that it has been renamed “Zif”. This matches with the <code>altname</code>, so we conclude that rows with a “CG” <code>tfid</code> can be safely skipped as their <code>altname</code> contains the new gene symbol.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">altname</span><span class="op">)</span> <span class="op">!=</span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">tfid</span><span class="op">)</span>,
                <span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"-|\\.|\\("</span>, <span class="va">altname</span><span class="op">)</span>,
                <span class="co"># Remove CG genes from consideration</span>
                <span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"CG\\d+"</span>, <span class="va">tfid</span><span class="op">)</span>
                <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">altname</span>, <span class="va">tfid</span>, <span class="va">name</span>, <span class="va">consensus</span><span class="op">)</span>
<span class="co">#&gt;    altname tfid                 name   consensus</span>
<span class="co">#&gt; 1      cyc  Clk     Clk_cyc_SANGER_5    MCACGTGA</span>
<span class="co">#&gt; 2   CG6272  Crc  Crc_CG6272_SANGER_5 ATTACRTCABC</span>
<span class="co">#&gt; 3      Max   dm     dm_Max_SANGER_10  RNCACGTGGT</span>
<span class="co">#&gt; 4      Clk  gce     gce_Clk_SANGER_5    GCCACGTG</span>
<span class="co">#&gt; 5      Jra  kay     kay_Jra_SANGER_5 NATGASTCAYC</span>
<span class="co">#&gt; 6  schlank Lag1            Lag1_Cell  CYACYAAAWT</span>
<span class="co">#&gt; 7  schlank Lag1          Lag1_SOLEXA  CYACYAAWWT</span>
<span class="co">#&gt; 8     Lim1  lim         lim_SOLEXA_2   NNTAATTRV</span>
<span class="co">#&gt; 9      Mnt  Max     Max_Mnt_SANGER_5     CCACGTG</span>
<span class="co">#&gt; 10     Clk  Met     Met_Clk_SANGER_5    GACACGTG</span>
<span class="co">#&gt; 11  bigmax  Mio  Mio_bigmax_SANGER_5    ATCACGTG</span>
<span class="co">#&gt; 12     Bgb  run          run_Bgb_NBT   WAACCGCAR</span>
<span class="co">#&gt; 13     Clk  tai     tai_Clk_SANGER_5   RCACGTGTC</span>
<span class="co">#&gt; 14     cyc  tgo     tgo_cyc_SANGER_5   GTCACGTGM</span>
<span class="co">#&gt; 15     sim  tgo     tgo_sim_SANGER_5  GWACGTGACC</span>
<span class="co">#&gt; 16    sima  tgo    tgo_sima_SANGER_5   GTACGTGAC</span>
<span class="co">#&gt; 17      ss  tgo      tgo_ss_SANGER_5    TGCGTGAC</span>
<span class="co">#&gt; 18     tai  tgo     tgo_tai_SANGER_5   RCACGTGAC</span>
<span class="co">#&gt; 19     trh  tgo     tgo_trh_SANGER_5  RTACGTGACC</span>
<span class="co">#&gt; 20  CG6272 Xrp1 Xrp1_CG6272_SANGER_5  ATTRCRYMAY</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [Note: incomplete universalmotif_df object.]</span></code></pre></div>
<p>The remaining rows (only 20 values) can be manually inspected for any discrepancies. I went through each entry by hand, looking up their motifs on <a href="https://mccb.umassmed.edu/ffs/" class="external-link">FlyFactor</a>, and their gene names on <a href="flybase.org">FlyBase</a> to determine the best way to handle these motifs. Sometimes the best way to be sure your data are high quality is to carefully inspect it!</p>
<p>I determined from this that a few <code>altnames</code> need swapping, and one motif I will remove because it is unusual (<a href="https://mccb.umassmed.edu/ffs/TFdetails.php?FlybaseID=FBgn0013753" class="external-link">Bgb</a> has an identical motif to <a href="https://mccb.umassmed.edu/ffs/TFdetails.php?FlybaseID=FBgn0003300" class="external-link">run</a>, but the motif is marked “run” on the FlyFactor website).</p>
<p>I’ll make those changes to the data:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">swap_alt_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CG6272"</span>, <span class="st">"Clk"</span>, <span class="st">"Max"</span>, <span class="st">"Mnt"</span>, <span class="st">"Jra"</span><span class="op">)</span>
<span class="va">remove</span> <span class="op">&lt;-</span> <span class="st">"Bgb"</span>

<span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>altname <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">altname</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">swap_alt_id</span>, <span class="va">tfid</span>, <span class="va">altname</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="op">(</span><span class="va">altname</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">remove</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Finally, the remaining motif metadata is also OK based on my manual inspection.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">altname</span><span class="op">)</span> <span class="op">!=</span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">tfid</span><span class="op">)</span>,
                <span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"-|\\.|\\("</span>, <span class="va">altname</span><span class="op">)</span>,
                <span class="co"># Remove CG genes from consideration</span>
                <span class="op">!</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"CG\\d+"</span>, <span class="va">tfid</span><span class="op">)</span>
                <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">altname</span>, <span class="va">tfid</span>, <span class="va">name</span>, <span class="va">consensus</span><span class="op">)</span>
<span class="co">#&gt;    altname tfid                name  consensus</span>
<span class="co">#&gt; 1      cyc  Clk    Clk_cyc_SANGER_5   MCACGTGA</span>
<span class="co">#&gt; 2  schlank Lag1           Lag1_Cell CYACYAAAWT</span>
<span class="co">#&gt; 3  schlank Lag1         Lag1_SOLEXA CYACYAAWWT</span>
<span class="co">#&gt; 4     Lim1  lim        lim_SOLEXA_2  NNTAATTRV</span>
<span class="co">#&gt; 5   bigmax  Mio Mio_bigmax_SANGER_5   ATCACGTG</span>
<span class="co">#&gt; 6      cyc  tgo    tgo_cyc_SANGER_5  GTCACGTGM</span>
<span class="co">#&gt; 7      sim  tgo    tgo_sim_SANGER_5 GWACGTGACC</span>
<span class="co">#&gt; 8     sima  tgo   tgo_sima_SANGER_5  GTACGTGAC</span>
<span class="co">#&gt; 9       ss  tgo     tgo_ss_SANGER_5   TGCGTGAC</span>
<span class="co">#&gt; 10     tai  tgo    tgo_tai_SANGER_5  RCACGTGAC</span>
<span class="co">#&gt; 11     trh  tgo    tgo_trh_SANGER_5 RTACGTGACC</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [Note: incomplete universalmotif_df object.]</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="removing-duplicate-motif-matrices">Removing duplicate motif matrices<a class="anchor" aria-label="anchor" href="#removing-duplicate-motif-matrices"></a>
</h2>
<p>Just because the metadata for each entry is unique, this does not mean that the motif matrix for each entry is unique. There are many reasons why two different factors could have identical motifs: some biological, others technical. In the case of the FlyFactorSurvey, some entries are duplicated in MotifDb which should not be.</p>
<p>For instance, the following motif is a duplicate where the tidied metadata matches:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">consensus</span> <span class="op">==</span> <span class="st">"MMCACCTGYYV"</span><span class="op">)</span>
<span class="co">#&gt;            motif altname               name   tfid      organism   consensus</span>
<span class="co">#&gt; 1 *     &lt;mot:da&gt;  HLH54F HLH54F_da_SANGER_5 HLH54F Dmelanogaster MMCACCTGYYV</span>
<span class="co">#&gt; 2 * &lt;mot:HLH54F&gt;  HLH54F HLH54F_da_SANGER_5 HLH54F Dmelanogaster MMCACCTGYYV</span>
<span class="co">#&gt;   alphabet strand  icscore nsites type pseudocount                    bkg</span>
<span class="co">#&gt; 1      DNA     +- 12.24152     23  PPM           1 0.25, 0.25, 0.25, 0.25</span>
<span class="co">#&gt; 2      DNA     +- 12.24152     23  PPM           1 0.25, 0.25, 0.25, 0.25</span>
<span class="co">#&gt;        dataSource</span>
<span class="co">#&gt; 1 FlyFactorSurvey</span>
<span class="co">#&gt; 2 FlyFactorSurvey</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [Hidden empty columns: family, bkgsites, pval, qval, eval.]</span>
<span class="co">#&gt; [Rows marked with * are changed. Run update_motifs() or to_list() to</span>
<span class="co">#&gt;   apply changes.]</span></code></pre></div>
<p>It is difficult to determine in a high-throughput way whether any matrix entries are identical in a large database, and it is not always possible to rely on metadata to determine matrix duplication.</p>
<p>In order to identify and remove duplicate motif matrices, memes provides <code><a href="../reference/remove_duplicate_motifs.html">remove_duplicate_motifs()</a></code>, which can be used to deduplicate a list of motifs based solely on their motif matrices (i.e. it ignores motif name &amp; other metadata). We will use this strategy to deduplicate the flyFactor data.</p>
<p>(NOTE: When working with other motif databases, it is critical to understand the data source to determine appropriate measures for handling duplicated entries.)</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># This operation takes a while to run on large motif lists</span>
<span class="va">flyFactor_dedup</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_duplicate_motifs.html">remove_duplicate_motifs</a></span><span class="op">(</span><span class="va">flyFactor_data</span><span class="op">)</span>
<span class="co">#&gt; Warning in universalmotif::compare_motifs(x, method = "PCC"): Some comparisons</span>
<span class="co">#&gt; failed due to low motif IC</span></code></pre></div>
<p>Duplicate removal identifies and removes 57 identical matrices.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Rows before cleanup</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">flyFactor_data</span><span class="op">)</span>
<span class="co">#&gt; [1] 613</span>
<span class="co"># Rows after cleanup</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">flyFactor_dedup</span><span class="op">)</span>
<span class="co">#&gt; [1] 556</span></code></pre></div>
<p>Using the example from before now shows only 1 motif corresponding to this sequence.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactor_dedup</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">consensus</span> <span class="op">==</span> <span class="st">"MMCACCTGYYV"</span><span class="op">)</span>
<span class="co">#&gt;          motif               name altname      organism   consensus alphabet</span>
<span class="co">#&gt; 1 &lt;mot:HLH5..&gt; HLH54F_da_SANGER_5  HLH54F Dmelanogaster MMCACCTGYYV      DNA</span>
<span class="co">#&gt;   strand  icscore nsites type pseudocount                    bkg   tfid</span>
<span class="co">#&gt; 1     +- 12.24152     23  PPM           1 0.25, 0.25, 0.25, 0.25 HLH54F</span>
<span class="co">#&gt;        dataSource</span>
<span class="co">#&gt; 1 FlyFactorSurvey</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [Hidden empty columns: family, bkgsites, pval, qval, eval.]</span></code></pre></div>
<p>Finally, now that the database has been tidied and deduplicated, the resulting data.frame can be converted back into a universalmotif list using <code><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">to_list()</a></code>. To discard the additional columns we created so they are not passed on to the <code>universalmotif</code>, set <code>extrainfo = FALSE</code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># extrainfo = FALSE drops the extra columns we added during data cleaning which are now unneeded</span>
<span class="va">flyFactorMotifs_final</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/tidy-motifs.html" class="external-link">to_list</a></span><span class="op">(</span><span class="va">flyFactor_dedup</span>, extrainfo <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Discarding unknown slot(s) 'tfid', 'dataSource' (set `extrainfo=TRUE`</span>
<span class="co">#&gt;   to preserve these).</span></code></pre></div>
<p>The resulting universalmotif list object now reflects the changes we made to the <code>data.frame</code> and can now be exported as a .meme format file using <code><a href="https://rdrr.io/pkg/universalmotif/man/write_meme.html" class="external-link">universalmotif::write_meme</a></code> or can be used directly as input to tools like <code>runTomTom</code> or <code>runAme</code>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">flyFactorMotifs_final</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;        Motif name:   ab_SANGER_10</span>
<span class="co">#&gt;    Alternate name:   ab</span>
<span class="co">#&gt;          Organism:   Dmelanogaster</span>
<span class="co">#&gt;          Alphabet:   DNA</span>
<span class="co">#&gt;              Type:   PPM</span>
<span class="co">#&gt;           Strands:   +-</span>
<span class="co">#&gt;          Total IC:   15.01</span>
<span class="co">#&gt;       Pseudocount:   1</span>
<span class="co">#&gt;         Consensus:   BWNRCCAGGWMCNNTNNNGNN</span>
<span class="co">#&gt;      Target sites:   20</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     B    W    N    R C C A G G    W    M    C    N    N    T    N    N    N</span>
<span class="co">#&gt; A 0.0 0.50 0.20 0.35 0 0 1 0 0 0.55 0.35 0.05 0.20 0.45 0.20 0.10 0.40 0.40</span>
<span class="co">#&gt; C 0.3 0.15 0.25 0.00 1 1 0 0 0 0.10 0.65 0.70 0.45 0.25 0.10 0.25 0.25 0.10</span>
<span class="co">#&gt; G 0.4 0.05 0.50 0.65 0 0 0 1 1 0.00 0.00 0.05 0.05 0.15 0.05 0.20 0.05 0.15</span>
<span class="co">#&gt; T 0.3 0.30 0.05 0.00 0 0 0 0 0 0.35 0.00 0.20 0.30 0.15 0.65 0.45 0.30 0.35</span>
<span class="co">#&gt;      G    N    N</span>
<span class="co">#&gt; A 0.25 0.50 0.30</span>
<span class="co">#&gt; C 0.10 0.25 0.25</span>
<span class="co">#&gt; G 0.55 0.15 0.45</span>
<span class="co">#&gt; T 0.10 0.10 0.00</span></code></pre></div>
<p>This cleaned-up version of the FlyFactorSurvey data is packaged with memes in <code>system.file("extdata/flyFactorSurvey_cleaned.meme", package = "memes")</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/universalmotif/man/write_meme.html" class="external-link">write_meme</a></span><span class="op">(</span><span class="va">flyFactorMotifs_final</span>, <span class="st">"flyFactorSurvey_cleaned.meme"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level1">
<h1 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h1>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.0 (2021-05-18)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.3 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] universalmotif_1.13.4 magrittr_2.0.1        memes_1.3.2          </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] MatrixGenerics_1.6.0        Biobase_2.54.0             </span>
<span class="co">#&gt;  [3] sass_0.4.0                  tidyr_1.1.4                </span>
<span class="co">#&gt;  [5] jsonlite_1.7.2              bslib_0.3.1                </span>
<span class="co">#&gt;  [7] stats4_4.1.0                GenomeInfoDbData_1.2.7     </span>
<span class="co">#&gt;  [9] Rsamtools_2.10.0            yaml_2.2.1                 </span>
<span class="co">#&gt; [11] pillar_1.6.4                lattice_0.20-44            </span>
<span class="co">#&gt; [13] glue_1.5.1                  digest_0.6.29              </span>
<span class="co">#&gt; [15] GenomicRanges_1.46.1        XVector_0.34.0             </span>
<span class="co">#&gt; [17] colorspace_2.0-2            htmltools_0.5.2            </span>
<span class="co">#&gt; [19] Matrix_1.3-4                XML_3.99-0.8               </span>
<span class="co">#&gt; [21] pkgconfig_2.0.3             ggseqlogo_0.1              </span>
<span class="co">#&gt; [23] zlibbioc_1.40.0             purrr_0.3.4                </span>
<span class="co">#&gt; [25] scales_1.1.1                tzdb_0.2.0                 </span>
<span class="co">#&gt; [27] BiocParallel_1.28.2         tibble_3.1.6               </span>
<span class="co">#&gt; [29] generics_0.1.1              IRanges_2.28.0             </span>
<span class="co">#&gt; [31] ggplot2_3.3.5               ellipsis_0.3.2             </span>
<span class="co">#&gt; [33] cachem_1.0.6                SummarizedExperiment_1.24.0</span>
<span class="co">#&gt; [35] BiocGenerics_0.40.0         splitstackshape_1.4.8      </span>
<span class="co">#&gt; [37] crayon_1.4.2                memoise_2.0.1              </span>
<span class="co">#&gt; [39] evaluate_0.14               fs_1.5.1                   </span>
<span class="co">#&gt; [41] fansi_0.5.0                 MASS_7.3-54                </span>
<span class="co">#&gt; [43] textshaping_0.3.6           data.table_1.14.2          </span>
<span class="co">#&gt; [45] tools_4.1.0                 hms_1.1.1                  </span>
<span class="co">#&gt; [47] BiocIO_1.4.0                lifecycle_1.0.1            </span>
<span class="co">#&gt; [49] matrixStats_0.61.0          stringr_1.4.0              </span>
<span class="co">#&gt; [51] S4Vectors_0.32.3            munsell_0.5.0              </span>
<span class="co">#&gt; [53] DelayedArray_0.20.0         Biostrings_2.62.0          </span>
<span class="co">#&gt; [55] compiler_4.1.0              pkgdown_2.0.1.9000         </span>
<span class="co">#&gt; [57] jquerylib_0.1.4             GenomeInfoDb_1.30.0        </span>
<span class="co">#&gt; [59] systemfonts_1.0.3           rlang_0.4.12               </span>
<span class="co">#&gt; [61] MotifDb_1.36.0              grid_4.1.0                 </span>
<span class="co">#&gt; [63] RCurl_1.98-1.5              rjson_0.2.20               </span>
<span class="co">#&gt; [65] bitops_1.0-7                rmarkdown_2.11             </span>
<span class="co">#&gt; [67] restfulr_0.0.13             gtable_0.3.0               </span>
<span class="co">#&gt; [69] R6_2.5.1                    GenomicAlignments_1.30.0   </span>
<span class="co">#&gt; [71] knitr_1.36                  dplyr_1.0.7                </span>
<span class="co">#&gt; [73] rtracklayer_1.54.0          fastmap_1.1.0              </span>
<span class="co">#&gt; [75] utf8_1.2.2                  rprojroot_2.0.2            </span>
<span class="co">#&gt; [77] ragg_1.2.0                  readr_2.1.1                </span>
<span class="co">#&gt; [79] desc_1.4.0                  stringi_1.7.6              </span>
<span class="co">#&gt; [81] parallel_4.1.0              Rcpp_1.0.7                 </span>
<span class="co">#&gt; [83] vctrs_0.3.8                 tidyselect_1.1.1           </span>
<span class="co">#&gt; [85] xfun_0.28</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Spencer Nystrom.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
